{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "R.version.string\n",
    "print(\"Finally, Bulk Typing of Bacterial Species down to Strain Level using ON-rep-seq\")\n",
    "print(\"Łukasz Krych, Josué L. Castro-Mejía, Daniel N. Moesby, Morten B. Mikkelsen, Morten A. Rasmussen, Maciej Sykulski, Dennis S. Nielsen\")\n",
    "print(\"-- \")\n",
    "print(\"https://bitbucket.org/modelscat/on-rep-seq/\")\n",
    "print(\"This notebook maintainer's email:\")\n",
    "print(\"Maciej Sykulski'<m.sykulski@genxone.eu>, 'Maciek Sykulski'<macieksk@gmail.com>\")\n",
    "length(ls())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#workingDir<-\"./\"; getwd();\n",
    "workingDir <- \"__WD_DIR\" \n",
    "setwd(workingDir); getwd()\n",
    "\n",
    "#projDir <- \"__PROJ_DIR\" \n",
    "projDir <- workingDir \n",
    "\n",
    "projectPrefix <- \"__PROJECT_PREFIX\"\n",
    "#projectPrefix <- \"runnable_jupyter_on-rep-seq_flowgrams_clustering_heatmaps\"\n",
    "if (projectPrefix == paste0(\"__\",\"PROJECT_PREFIX\")) {\n",
    "    projectPrefix <- \"runnable_jupyter_on-rep-seq_flowgrams_clustering_heatmaps\" }\n",
    "\n",
    "\n",
    "## To obtain publication .Rdata contact with the authors\n",
    "\n",
    "####### !!!!!!!!\n",
    "####### UNCOMMENT below line and execute this cell\n",
    "####### to load the executed notebook .Rdata to continue to work with it\n",
    "####### !!!!!!!!\n",
    "\n",
    "#LOAD_SAVED_IMAGE<-TRUE\n",
    "\n",
    "if (exists(\"LOAD_SAVED_IMAGE\") && LOAD_SAVED_IMAGE) {\n",
    "  system.time(load(paste(projDir,\"/r_saved_images/\",projectPrefix,\".Rdata\",sep=\"\")))\n",
    "  printS<-print\n",
    "  load.install.libraries()\n",
    "  lsos()\n",
    "}\n",
    "    \n",
    "gc()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "printS<-print\n",
    "fileStdOut<-\"__STDOUT_FILE\"\n",
    "if(fileStdOut!=paste0(\"__\",\"STDOUT_FILE\")){\n",
    "    fileConn<-file(fileStdOut, \"w\")\n",
    "    #fileConn<-open(fileStdOut, open = \"w\", blocking = FALSE)\n",
    "    #writeLines(c(\"Hello\",\"World\"), fileConn)\n",
    "    printS<-function(obj,...){\n",
    "        text<-capture.output(print(obj))\n",
    "        writeLines(text, fileConn); flush(fileConn)\n",
    "        print(obj,...)\n",
    "    }\n",
    "    printS(paste0(\"Using \",fileStdOut,\" as the output sink.\"))\n",
    "}  else {printS<-print}\n",
    "#close(fileConn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "srcDir <- \"__SRC_PARENT_DIR\"\n",
    "#testvar <- \"__TEST_VAR\"\n",
    "\n",
    "rep.sample.dir=\"__REP_SAMPLE_DIR\" #paste0(projDir,\"/../data/Lukas180425/peaks-profiles/\")\n",
    "#rep.sample.dir=\"paste0(projDir,\"/../data/Lukas180425/peaks-profiles/\")\n",
    "\n",
    "ma.window.size <- as.integer(\"__MA_WINDOW_SIZE\")\n",
    "if(is.na(ma.window.size)) ma.window.size <- 20 ## default value\n",
    "printS(paste0(\"Using ma.window.size=\",ma.window.size))\n",
    "\n",
    "heatmap.filepref <- \"__HMAP_PREFIX\"\n",
    "if (heatmap.filepref == paste0(\"__\",\"HMAP_PREFIX\")) {\n",
    "    heatmap.filepref <- \"on-rep-seq\" }\n",
    "heatmap.filepref <- paste0(projDir,\"/\",heatmap.filepref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "raises-exception"
    ]
   },
   "outputs": [],
   "source": [
    "##This cell is marked with 'raises-exception' tag (View/Cell Toolbar/Tags) so the notebook continues - in case if not allowing exceptions in the notebook, which currently is not a good idea since in case of error the execution resulting notebook is not saved. Parameter --save-on-error for jupyter nbconvert is not yet implemented.\n",
    "stop(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "stopNbEval<-function(...){\n",
    "    defname <- paste(projDir, \"/r_saved_images/\", projectPrefix, \".Rdata\", sep = \"\")\n",
    "    printS(p(\"Stopping Notebook evaluation effectively by deleting objects. Saving first R image .Rdata to: \", defname));\n",
    "    printS(head(lsos(),3)); printS(gc());\n",
    "    save_new_image();\n",
    "    #i<-1; while(TRUE){\n",
    "    for (i in c(1)) {                          \n",
    "     rm(pos=i,list = ls(pos=i))\n",
    "     #tryCatch(rm(pos=i,list = ls(pos=i,all.names = TRUE)),error=function(e){break})\n",
    "};}#i<-i+1;}; stop(1)}\n",
    "#stopNbEval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#srcDir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryI<-function(pkg){pkg<-as.character(substitute(pkg)); print(pkg);\n",
    "if(!library(pkg,logical.return=TRUE,character.only=TRUE)){\n",
    "    install.packages(pkg);\n",
    "    if(!library(pkg,logical.return=TRUE,character.only=TRUE)){\n",
    "        printS(p(\"ERROR! could not load package \",pkg)); stop(1)\n",
    "}}}\n",
    "#libraryI(latticeee)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraryI(heatmap3)\n",
    "libraryI(hash)\n",
    "libraryI(fastmatch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "load.sources<-function(){\n",
    "    for (f in  c(\"utils.R\", #Contains inpar, save_new_image, etc\n",
    "                 \"heatmap3_env.R\" #The version of heatmap3 used in the project is slightly different (older?) than the one from heatmap3 package\n",
    "                 #...  \n",
    "                 ) \n",
    "        ){\n",
    "    source(paste(srcDir,\"/src/\",f,sep=\"\"))\n",
    "   }\n",
    "}\n",
    "load.sources()\n",
    "\n",
    "psz<-function(w,h=w,res=120,...)options(repr.plot.width=w,  repr.plot.height=h,repr.plot.res=res,...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gc()\n",
    "#system.time(save_new_image())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load.install.libraries<-function(){\n",
    "libraryI(parallel)\n",
    "libraryI(lattice)\n",
    "libraryI(latticeExtra)\n",
    "libraryI(ggplot2)\n",
    "libraryI(gplots)\n",
    "#libraryI(png)\n",
    "\n",
    "#libraryI(Hmisc)\n",
    "libraryI(cluster)\n",
    "libraryI(MASS)\n",
    "libraryI(reshape)\n",
    "libraryI(heatmap3)\n",
    "##libraryI(forecast) ## ma moving average, \n",
    "## using ma defined few cells down\n",
    "#ma<-function (x, n = 5) { filter(x, rep(1/n, n), sides = 2) }\n",
    "\n",
    "#libraryI(lumi)\n",
    "#libraryI(lumiHumanIDMapping)\n",
    "#libraryI(lumiHumanAll.db)\n",
    "#libraryI(limma)\n",
    "\n",
    "#libraryI(annotate)\n",
    "#libraryI(GOstats)\n",
    "\n",
    "#libraryI(multtest)\n",
    "libraryI(matrixStats)\n",
    "#libraryI(extrafont)\n",
    "#libraryI(enrichvs)\n",
    "#libraryI(statmod)\n",
    "#libraryI(flexmix)\n",
    "\n",
    "libraryI(gdata)  ## read.xls\n",
    "\n",
    "libraryI(IRdisplay)\n",
    "libraryI(scales)\n",
    "\n",
    "libraryI(spectral)\n",
    "\n",
    "load.sources()\n",
    "}\n",
    "\n",
    "load.install.libraries()\n",
    "\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#install.packages(\"heatmap3\")\n",
    "#for (i in 1:1000){dev.off()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#library(scales)\n",
    "#grey_pal\n",
    "\n",
    "ma<-function (x, n = 5) { filter(x, rep(1/n, n), sides = 2) }\n",
    "\n",
    "plot.matrix<-function(z,low=0,high=1){    \n",
    "    image(1:ncol(z), 1:nrow(z), round(t((z-low)/(high-low))*100), \n",
    "          #col = grey_pal(1,0)(100),\n",
    "          col = c(grey_pal(1,0.5)(20),grey_pal(0.5,0)(80)), \n",
    "          axes = FALSE)\n",
    "}\n",
    "\n",
    "plot.matrix.fp<-function(z,low=min(z,na.rm=TRUE),high=max(z,na.rm=TRUE), center=F){\n",
    "   z<-as.matrix(z)\n",
    "   if(center) z <- z-mean(z)\n",
    "   image(1:ncol(z), 1:nrow(z), t(z), #round(t((z-low)/(high-low))*100), \n",
    "         #col = colors()[1:100],\n",
    "         #col = grey_pal(0,1)(100),\n",
    "         col = greenred(100),\n",
    "         #col = c(grey_pal(1,0.5)(20),grey_pal(0.5,0)(80)), \n",
    "         axes = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "# <font size=\"10px\"><center> Developing pipeline for Oxford Nanopore based rep-PCR amplicon sequencing </center></font>\n",
    "\n",
    "![EM_title_part](EM_title_part.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fnames<-sort(dir(p(projDir,\"/../porechop_out.ONTsmth_GTGrem_lenstats/\"),\"*txt\"))\n",
    "\n",
    "#rep.sample.dir<-p(projDir,\"/../data/Lukas180425/peaks-profiles/\")\n",
    "fnames<-grep(\"^rep\",sort(dir(rep.sample.dir,\"*txt\")),value=TRUE)\n",
    "\n",
    "printS(p(\"Found \",length(fnames),\" rep*.txt files in the sample dir=\",rep.sample.dir))\n",
    "\n",
    "printS(p(\"head(fnames)=\",p(head(fnames),collapse=\", \")))\n",
    "printS(p(\"table(table(fnames))=\",table(table(fnames))))\n",
    "#substr(fnames,1,13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Possible TODO support sample description in .xls\n",
    "if (exists(\"SAMPLES_NAMES_XLS\") && !is.null(SAMPLES_NAMES_XLS)) {\n",
    "  libraryI(gdata)\n",
    "  rep.assign.wide<-read.xls(\n",
    "  #p(projDir,\"/../data/Lukas180425/peaks-profiles/Table01.xls\"),\n",
    "  #p(rep.sample.dir,\"/Table01.xls\"),\n",
    "  p(SAMPLES_NAMES_XLS),      \n",
    "  method=\"tsv\")\n",
    "    \n",
    "  head(rep.assign.wide)\n",
    "  table(rep.assign.wide$id)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "! [EM_peaks_all](EM_peaks_all.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#samp.len.cnts<-data.frame()\n",
    "maxL<-8000\n",
    "samp.len.cnts<-matrix(0,maxL,length(fnames))\n",
    "#dim(samp.len.cnts)\n",
    "for (i in seq_along(fnames)){\n",
    "    f<-fnames[i]\n",
    "    cnts<-read.table(paste(rep.sample.dir,f,sep=\"/\"),header=FALSE)\n",
    "    cnts[,1]<-as.integer(cnts[,1])\n",
    "    cnts[,2]<-as.integer(cnts[,2])\n",
    "    #print(class(cnts[,1]))\n",
    "    cnts<-cnts[cnts[,2]<=maxL,] #Subset only these lengths\n",
    "    samp.len.cnts[cnts[,2],i]<-cnts[,1]            \n",
    "}\n",
    "colnames(samp.len.cnts)<-sub(\".txt$\",\"\",fnames)\n",
    "colnames(samp.len.cnts)<-sub(\".fastq$\",\"\",colnames(samp.len.cnts))\n",
    "options(repr.plot.width=10, repr.plot.height=4,res=120)\n",
    "plot(sort(colSums(samp.len.cnts)),main=\"Read counts within barcoded samples of one 4hr MinION run\",\n",
    "     xlab=\"Barcoded subsample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "which.max(colSums(samp.len.cnts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Possible TODO support sample description in .xls\n",
    "if (exists(\"SAMPLES_NAMES_XLS\") && !is.null(SAMPLES_NAMES_XLS)) {\n",
    "rep.assign<-reshape(rep.assign.wide,direction = \"long\", varying = c('run08A','run08B','run08C'), times=c(\"A\",\"B\",\"C\"),\n",
    "        v.names = \"rep.name.pref\",new.row.names=1:99999)\n",
    "rep.assign<-transform(rep.assign,rep.name=paste(rep.name.pref,time,sep=\"\"),\n",
    "                      strain=gsub(\" \",\"_\",trimws(strain)))\n",
    "rep.assign<-transform(rep.assign,\n",
    "                      full.strain=gsub(\" \",\"_\",paste(strain,type.strain, sep=\"_\")))\n",
    "table(table(rep.assign$rep.name))\n",
    "head(rep.assign)\n",
    "} else {\n",
    "  ## This works now: sample names and strains from filenames, \n",
    "  ## no a priori clustering\n",
    "  vnms<-colnames(samp.len.cnts) ## Has to be exact! is matched later\n",
    "  rep.assign<-data.frame(id=vnms,\n",
    "    plate.format=vnms,\n",
    "    strain=vnms,\n",
    "    type.strain=vnms,\n",
    "    profile.ID=vnms,\n",
    "    ID=vnms,\n",
    "    time=vnms,\n",
    "    rep.name.pref=vnms,\n",
    "    rep.name=vnms,\n",
    "    full.strain=vnms\n",
    "    )    \n",
    "}    \n",
    "\n",
    "table(table(with(rep.assign,strain)))\n",
    "table(table(with(rep.assign,full.strain)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s.num<-1\n",
    "s.num<-NCOL(samp.len.cnts)\n",
    "options(repr.plot.width=10, repr.plot.height=3,res=120)\n",
    "plot(samp.len.cnts[1:5000,s.num],type=\"l\",pch=\".\",xlab=\"read.length\",main=\"A single bacterial colony sample profile\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s.num<-1\n",
    "s.num<-NCOL(samp.len.cnts)\n",
    "options(repr.plot.width=10, repr.plot.height=3,res=120)\n",
    "plot(x=log2(30:5000),y=samp.len.cnts[30:5000,s.num],type=\"l\",pch=\".\",xlab=\"log2(read.length)\",main=\"A single bacterial colony sample profile (log2 xscale)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save_new_image()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "printS(\"Measuring jitteriness of profiles\")\n",
    "cnt.range<-150:3000\n",
    "\n",
    "jitter.res<-mclapply(10:80,function(ma.window){\n",
    "samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)as.integer(ma(r,ma.window)))\n",
    "samp.prob<-samp.len.cnts.ma[cnt.range,]\n",
    "samp.prob<-t(t(samp.prob)/(colSums(samp.prob)+1))\n",
    "mean(sapply(1:NCOL(samp.prob),function(i)sum(abs(diff(sign(diff(samp.prob[,i])))))/2))\n",
    "})  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Proposed ma.window.size TODO make variable DONE\n",
    "#ma.window.size<-20 ## Now defined at the top\n",
    "\n",
    "#jitter.res\n",
    "printS(\"Jitteriness plots\")\n",
    "psz(h=6,w=7)\n",
    "par(mfcol=c(2,1), mai = c(0.8, 0.8, 0.5, 0.1),oma=c(0,0,0,0))\n",
    "layout(matrix(c(1,2),2, 1, byrow = TRUE), \n",
    "  \twidths=c(1,1), heights=c(1.07,0.93))\n",
    "plot(10:80,jitter.res,type=\"b\",ylab=\"mean.jitter of profiles\",xlab=\"ma.window\")\n",
    "title(main=\"Choice of ma window size based on profiles jitter\")\n",
    "grid()\n",
    "abline(v=ma.window.size,lty=2,col=\"darkgrey\")\n",
    "par(mai=c(0.8, 0.8, 0.1, 0.1))\n",
    "plot(11:80,diff(as.numeric(jitter.res)),type=\"b\",ylab=\"diff.mean.jitter of profiles\",xlab=\"ma.window\")\n",
    "grid()\n",
    "\n",
    "abline(v=ma.window.size,lty=2,col=\"darkgrey\")\n",
    "#title(\"\",line=-1)\n",
    "#psz(h=4,w=10)\n",
    "#plot(12:80,diff(diff(as.numeric(jitter.res),lag=1)),type=\"b\",ylab=\"diff.diff.mean.jitter of profiles\",xlab=\"ma.window\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(h=4,w=10)\n",
    "#plot(11:80,diff(as.numeric(jitter.res)),type=\"b\",ylab=\"diff.mean.jitter of profiles\",xlab=\"ma.window\")\n",
    "plot(12:80,diff(diff(as.numeric(jitter.res),lag=1)),type=\"b\",ylab=\"diff.diff.mean.jitter of profiles\",xlab=\"ma.window\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#jitter.res\n",
    "#diff(as.numeric(jitter.res))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samp.len.cnts.org<-samp.len.cnts\n",
    "\n",
    "if (exists(\"SUBSAMPLE_FRAC\") && !is.null(SUBSAMPLE_FRAC)) {\n",
    "  ##Subsample \n",
    "  l<-1:NROW(samp.len.cnts.org)\n",
    "  frac<-SUBSAMPLE_FRAC #0.1\n",
    "  samp.len.cnts<-sapply(1:NCOL(samp.len.cnts.org),function(i){\n",
    "    src<-rep(l,samp.len.cnts.org[,i])\n",
    "    tabulate(sample(src,length(src)*frac),NROW(samp.len.cnts.org))\n",
    "  })\n",
    "  dim(samp.len.cnts)\n",
    "}\n",
    "##Use a whole sample \n",
    "#samp.len.cnts<-samp.len.cnts.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(colSums(samp.len.cnts)/colSums(samp.len.cnts.org))\n",
    "summary(colSums(samp.len.cnts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s.num<-1\n",
    "s.num<-NCOL(samp.len.cnts)\n",
    "\n",
    "##ma defined previously\n",
    "#ma <- function(x,n=5){filter(x,rep(1/n,n), sides=2)}\n",
    "#s<-sample(10,10,replace=TRUE)\n",
    "#rbind(s,as.integer(ma(s)),ma(s))\n",
    "\n",
    "#samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)as.integer(ma(r,20)))\n",
    "samp.len.cnts.ma<-apply(samp.len.cnts,2, function(r)ma(r,ma.window.size))\n",
    "options(repr.plot.width=10, repr.plot.height=3,res=120)    \n",
    "plot(samp.len.cnts.ma[1:5000,s.num],type=\"l\",pch=\".\",xlab=\"read.length\",main=p(\"A single bacterial colony sample profile\\n(MA \",ma.window.size,\"bp smoothed)\")) \n",
    "plot(x=log2(30:5000),y=log2(1+samp.len.cnts.ma[30:5000,s.num]),type=\"l\",pch=\".\",xlab=\"log2(read.length)\",main=p(\"A single bacterial colony sample profile\\n(MA \",ma.window.size,\"bp smoothed, log2 xscale)\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#s<-samp.len.cnts[400:500]\n",
    "#rbind(s,ma(s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#install.packages(\"forecast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20,res=120)\n",
    "printS(\"LCprofiles for the first 20 samples\")\n",
    "par(mfrow=c(5,4))\n",
    "for (i in 1:20){\n",
    " plot(samp.len.cnts.ma[1:3000,i],type=\"l\",pch=\".\")     \n",
    "}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sample reads (lengths) from mixture of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "<br>\n",
    "<br>\n",
    "\n",
    "# <center> Clustering Length Counts profiles </center>\n",
    "<br>\n",
    "<br>\n",
    "<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##\n",
    "## Normalize database of known profiles\n",
    "##\n",
    "\n",
    "samp.len.cnts.ma.norm<-apply(samp.len.cnts.ma,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "samp.len.cnts.ma.norm[is.na(samp.len.cnts.ma.norm)]<-0\n",
    "\n",
    "samp.len.cnts.norm<-apply(samp.len.cnts,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "samp.len.cnts.norm[is.na(samp.len.cnts.norm)]<-0                             \n",
    "                             \n",
    "\n",
    "all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-11)\n",
    "all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#samp.len.cnts samp.len.cnts.ma\n",
    "\n",
    "h0.sample<-rowSums(samp.len.cnts.norm)/sum(samp.len.cnts.norm)\n",
    "#h0.sample<-rowSums(samp.len.cnts)/sum(samp.len.cnts)\n",
    "\n",
    "sum(h0.sample)\n",
    "\n",
    "#options(repr.plot.width=10, repr.plot.height=3,res=120)\n",
    "psz(10,2.5,res=150); #par(mfcol=c(2,1))\n",
    "plot(h0.sample,type=\"l\")\n",
    "plot(h0.sample[150:3000],type=\"l\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "summary(colSums(samp.len.cnts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare.h0.odds.ratio.mat<-function(cnt.range=150:3000,noise.mixture.p=0.01){\n",
    "    cnt.range<<-cnt.range\n",
    "\n",
    "    #samp.len.cnts samp.len.cnts.ma\n",
    "    h0.sample<-rowSums(samp.len.cnts.norm)/sum(samp.len.cnts.norm)\n",
    "    #h0.sample<-rowSums(samp.len.cnts)/sum(samp.len.cnts)\n",
    "    sum(h0.sample)\n",
    "    #options(repr.plot.width=10, repr.plot.height=3,res=120)\n",
    "    #plot(h0.sample,type=\"l\")\n",
    "\n",
    "    uniform.sample<-rep(1/length(cnt.range),uniform=NROW(samp.len.cnts.ma.norm))\n",
    "    #h0.sample<-uniform.sample\n",
    "\n",
    "    #summary(samp.len.cnts.ma.norm[1000:2000,])\n",
    "\n",
    "    samp.prob<-cbind(h0.sample,uniform.sample,\n",
    "                                samp.len.cnts.ma.norm)[cnt.range,]\n",
    "    samp.prob<-t(t(samp.prob)/colSums(samp.prob))\n",
    "    #histogram(log10(as.vector(samp.prob[samp.prob!=0])),breaks=40)\n",
    "    r<-noise.mixture.p #0.01\n",
    "    noise.prob<-r/(1-r)/length(cnt.range)\n",
    "    c(np=noise.prob,qp=quantile(samp.prob[samp.prob!=0],noise.mixture.p)) #0.01))\n",
    "    #noise.prob<-quantile(samp.prob[samp.prob!=0],0.01)\n",
    "\n",
    "    samp.prob<-samp.prob+noise.prob\n",
    "    samp.prob<-t(t(samp.prob)/colSums(samp.prob))\n",
    "    stopifnot(all(abs(colSums(samp.prob)-1)<10^-8))\n",
    "    log.samp.prob<-log10(samp.prob)\n",
    "    \n",
    "    #print(table(colMins(log.samp.prob)))\n",
    "    print(head(table(colMins(log.samp.prob)),n=3))\n",
    "    \n",
    "    print(c(class(log.samp.prob),dim(log.samp.prob)))\n",
    "    print(c(class(samp.len.cnts),dim(samp.len.cnts)))\n",
    "\n",
    "    res<-mclapply(as.data.frame(#cbind(cbind(h0.sample,uniform.sample)*median(colSums(samp.len.cnts)),\n",
    "                                        samp.len.cnts),function(slcnts){\n",
    "                  #print(length(slcnts))\n",
    "                  #print(length(cnt.range))\n",
    "                  #stopifnot(length(slcnts[cnt.range])==NROW(log.samp.prob))\n",
    "                  colSums(log.samp.prob*slcnts[cnt.range])/sum(slcnts[cnt.range])\n",
    "                  }\n",
    "            #SIMPLIFY=TRUE\n",
    "            )\n",
    "    res<-as.data.frame(res)\n",
    "    print(c(\"NCOL(log.samp.prob)\"=NCOL(log.samp.prob)))\n",
    "    print(c(\"NCOL(samp.len.cnts)\"=NCOL(samp.len.cnts)))\n",
    "    print(dim(res))\n",
    "    print(class(res))\n",
    "\n",
    "    #h0.odds.ratio.mat<-(as.matrix(res[-(1:2),-(1:2)]-as.vector(as.matrix(res[-(1:2),1]))))#+ #column substraction\n",
    "    #h0.odds.ratio.mat<- as.matrix(t(res[-(1:2),-(1:2)])-as.vector(as.matrix(res[1,-(1:2)]))) #)/2\n",
    "\n",
    "    ##Log odds ratio according to above mclapply\n",
    "    #h0.odds.ratio.mat<- as.matrix(t(res[-(1:2),])-as.vector(as.matrix(res[1,]))) #)/2  \n",
    "    ##h0.sample\n",
    "    #h0.odds.ratio.mat<- as.matrix(t(res[-(1:2),])-as.vector(as.matrix(res[2,]))) #)/2  ##uniform.sample\n",
    "\n",
    "    ##Relative entropy, that is Kullback-Leiber divergence\n",
    "    h0.odds.ratio.mat<- as.matrix(t(res[-(1:2),])-as.vector(diag(as.matrix(res[-c(1:2),])))) #)/2  \n",
    "\n",
    "    #h0.odds.ratio.mat<-\n",
    "    #colSums(samp.len.cnts[cnt.range,])[sub.range])\n",
    "\n",
    "    ### Symmetrize\n",
    "    h0.odds.ratio.mat<-(h0.odds.ratio.mat+t(h0.odds.ratio.mat))/2\n",
    "\n",
    "    #unif.odds.ratio.mat<-TODO as above\n",
    "\n",
    "    print(dim(h0.odds.ratio.mat))\n",
    "    printS(\"summary(as.vector(h0.odds.ratio.mat))\")\n",
    "    printS(summary(as.vector(h0.odds.ratio.mat)))\n",
    "    h0.odds.ratio.mat\n",
    "}\n",
    "\n",
    "h0.odds.ratio.mat<-prepare.h0.odds.ratio.mat()\n",
    "#plot(sort(res[])\n",
    "#         ,decreasing=TRUE)[1:30])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#colSums(10^log.samp.prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(diag(h0.odds.ratio.mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#res[3,]\n",
    "#summary(t(res[3,]-res[2,]))\n",
    "\n",
    "psz(6)\n",
    "plot.matrix.fp(h0.odds.ratio.mat,center=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(8,3)\n",
    "#plot(sort(res[-(1:2),]-res[-(1:2),1],decreasing=TRUE)[1:30])\n",
    "samp.n<-15\n",
    "plot(sort(h0.odds.ratio.mat[,samp.n]\n",
    "          #/sum(samp.len.cnts[cnt.range,samp.n])          \n",
    "          ,decreasing=TRUE)[1:30],\n",
    "     ylab=\"scores\",\n",
    "     type=\"b\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#as.vector(as.matrix(res[1,-(1:2)]))\n",
    "rep.assign<-rep.assign[match(colnames(h0.odds.ratio.mat),rep.assign$rep.name),]\n",
    "stopifnot(NROW(rep.assign)==NCOL(samp.len.cnts))\n",
    "\n",
    "#TO DONE add read counts\n",
    "rep.assign$read.count<-colSums(samp.len.cnts[,rep.assign$rep.name])\n",
    "head(rep.assign)\n",
    "\n",
    "psz(3)\n",
    "hist(rep.assign$read.count,breaks=50)\n",
    "\n",
    "#TODO check A or B or C seperately? maybe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(as.vector(as.matrix(h0.odds.ratio.mat)))\n",
    "length(as.matrix(h0.odds.ratio.mat))\n",
    "table(as.integer(factor(rep.assign$full.strain[1:60])))\n",
    "#hc<-fastcluster::hclust(as.dist(-as.matrix(h0.odds.ratio.mat)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare.sub.data.to.heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# from https://gist.github.com/Jfortin1/72ef064469d1703c6b30\n",
    "darken <- function(color, factor=1.4){\n",
    "    col <- col2rgb(color)\n",
    "    col <- col/factor\n",
    "    col <- rgb(t(col), maxColorValue=255)\n",
    "    col\n",
    "}\n",
    "lighten <- function(color, factor=1.4){\n",
    "    col <- col2rgb(color)\n",
    "    col <- col*factor\n",
    "    col <- rgb(t(col), maxColorValue=255)\n",
    "    col\n",
    "}\n",
    "\n",
    "prepare.sub.data.to.heatmap<-function(sub.range){\n",
    "samp.labs<<-as.character(as.vector(rep.assign$.strain.[sub.range])) #was full.strain\n",
    "samp.color.labs<<-as.character(as.vector(rep.assign$.strain.similar[sub.range]))\n",
    "#print(t(t(table(samp.labs))))\n",
    "sl.stat<-t(t(table(table(samp.labs))))\n",
    "colnames(sl.stat)<-\"Number of a priori clusters based on `samp.labs` with a given size\"\n",
    "print(sl.stat)\n",
    "\n",
    "my.colors<<-rainbow(length(levels(factor(samp.labs))))\n",
    "my.colors<<-cbind(my.colors,darken(my.colors))\n",
    "#dim(my.colors)\n",
    "\n",
    "toplot.mat<<-h0.odds.ratio.mat[sub.range,sub.range]\n",
    "\n",
    "#toplot.mat<-t(t(toplot.mat)/colSums(samp.len.cnts[cnt.range,])[sub.range])\n",
    "\n",
    "#colnames(toplot.mat)<-samp.labs\n",
    "rownames(toplot.mat)<<-samp.labs\n",
    "    \n",
    "print(paste(\"max(toplot.mat)=\",max(toplot.mat)))\n",
    "\n",
    "#dist.matrix<<- 1-as.matrix((function(m){m/max(m)})(toplot.mat-min(toplot.mat)),\n",
    "                                #h0.odds.ratio.mat[sub.range,sub.range],\n",
    "                                #t(t(h0.odds.ratio.mat[sub.range,sub.range])/colSums(samp.len.cnts[cnt.range,]))\n",
    "#                                    )\n",
    "#dist.matrix<<- -as.matrix(toplot.mat-max(toplot.mat))\n",
    "#dist.matrix<<- -as.matrix(toplot.mat-0.6)\n",
    "dist.matrix<<- -as.matrix(toplot.mat)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(rep.assign$type.strain)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## rep.assign FIX thetaiotamicron_ATCC29148 label\n",
    "#ids.tofix<-rep.assign$rep.name%in%c(\"repBC71A\",\"repBC72B\",\"repBC65C\",\"repBC79A\",\"repBC80B\",\"repBC73C\")\n",
    "#rep.assign$full.strain<-as.character(rep.assign$full.strain)\n",
    "#rep.assign$full.strain[ids.tofix]<-#'Bacteroides_thetaiotaomicron_ATCC29148'\n",
    "#rep.assign$full.strain<-factor(rep.assign$full.strain)\n",
    "#rep.assign$strain<-as.character(rep.assign$strain)\n",
    "#rep.assign$strain[ids.tofix]<-'Bacteroides_thetaiotaomicron'\n",
    "#rep.assign$strain<-factor(rep.assign$strain)\n",
    "#rep.assign$type.strain<-as.character(rep.assign$type.strain)\n",
    "#rep.assign$type.strain[ids.tofix]<-'ATCC29148'\n",
    "#rep.assign$type.strain<-factor(rep.assign$type.strain)\n",
    "#rep.assign[ids.tofix,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "head(rep.assign)\n",
    "levels(rep.assign$full.strain)\n",
    "\n",
    "##Below lists not used in thii script\n",
    "\n",
    "##Rename all from the list to the first element in the list\n",
    "#rename.list<-list(\n",
    "#c('Akkermansia_muciniphila_DSMZ_22959', 'Akkermansia_muciniphila', 'Akkermansia_muciniphila_'),\n",
    "#c('Bacillus_cereus_15', 'Bacillus_cereus_15'),\n",
    "#c('Bacillus_cereus_GR177', 'Bacillus_cereus_38', 'Bacillus_cereus_38_(GR177)'),\n",
    "#[...]\n",
    "#)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rename.list<-list() ##Above list not used\n",
    "\n",
    "rep.assign$.strain.<-rep.assign$full.strain\n",
    "for (v in rename.list) {\n",
    "  to.ren<-levels(rep.assign$.strain.)%in%v\n",
    "  levels(rep.assign$.strain.)[to.ren]<-v[1]\n",
    "}\n",
    "table(rep.assign$.strain.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#similar.labels<-list(\n",
    " #   c(\"S.enterica_Typhimurium_u292\",\"S.enterica_Typhimurium_C5\"),\n",
    " #   c(\"Listeria_monocytogenes_EGDe\",\"Listeria_monocytogenes_LO28\"),\n",
    " #   c(\"Lactococcus_lactis_spp._cremoris_MG1363\",\"Lactococcus_lactis_spp._cremoris_Wg2\"),\n",
    "#    c(\"Lactobacillus_plantarum_LLFH15\",\"Lactobacillus_plantarum_DSM20174\"),\n",
    "#    c(\"Lactobacillus_paracasei_subsp._paracasei_NCFB151T\",\"Lactobacillus_paracasei_subsp._paracasei_own\")\n",
    "#    ,c('Bacteriodes_thethaiotaomicron_DSM2079','Bacteroides_thetaiotaomicron_ATCC29148')\n",
    "#)                     \n",
    "#sapply(1:length(similar.labels),function(i)sum(rep.assign$.strain.%in%similar.labels[[i]]))\n",
    "\n",
    "##Above list not used       \n",
    "similar.labels<-list()\n",
    "       \n",
    "       \n",
    "rep.assign$.strain.similar<-rep.assign$.strain.       \n",
    "       \n",
    "for (v in similar.labels) {\n",
    "  to.ren<-levels(rep.assign$.strain.similar)%in%v\n",
    "  levels(rep.assign$.strain.similar)[to.ren]<-v[1]\n",
    "}\n",
    "\n",
    "#t(t(table(rep.assign$.strain.similar)))\n",
    "t(t(table(table(rep.assign$.strain.similar))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#rep.assign[rep.assign$rep.name%in%c(\"repBC68A\",\"repBC69B\",\"repBC70C\",\"repBC76A\",\"repBC77B\",\"repBC78C\",\"repBC71A\",\"repBC72B\",\"repBC65C\",\"repBC79A\",\"repBC80B\",\"repBC73C\"),]\n",
    "#\"Bacteriodes_thethaiotaomicron_DSM2079\"==\"Bacteroides_thethaiotaomicron_DSM2079\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sort(rownames(t(t(table(rep.assign$.strain.)))))\n",
    "#t(t(table(rep.assign$.strain.)))\n",
    "t(t(table(table(rep.assign$.strain.))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#png(\"/dev/null\")\n",
    "#plot(1:100)\n",
    "#dev.off()\n",
    "##plot(1:10)\n",
    "\n",
    "#mean(grepl(\"C$\",as.character(rep.assign$rep.name)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "printS(\"LCprofiles heatmap\")\n",
    "suppressWarnings(\n",
    "sub.range<-which(grepl(#\"Salmonella\",\n",
    "                       #\"Bacillus|Salmonella|Listeria\",\n",
    "                       \"^rep\",\n",
    "                       rep.assign$full.strain)\n",
    "                 #&(!grepl(#\"Salmonella\",\n",
    "                 #      \"subtilis|lichniformis\",\n",
    "                 #      rep.assign$full.strain))\n",
    "                 #&(!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    ")\n",
    "#TODO\n",
    "#Subtilis Lihniformis out\n",
    "\n",
    "\n",
    "#\"Salmonella_typhimurium_u292\"\n",
    "#sub.range<-1:288\n",
    "\n",
    "prepare.sub.data.to.heatmap(sub.range)\n",
    "\n",
    "fslabs<-factor(samp.labs)\n",
    "#levels(fslabs)[grepl(\"Bacillus_cereus_38\",levels(fslabs))]<-\"Bacillus_cereus_38\"\n",
    "color.labs<-as.integer(fslabs)\n",
    "\n",
    "\n",
    "#heatmap3(res.log.scaled.sparse$L.svd$vt[1:8,] * res.log.scaled.sparse$L.svd$d[1:8], \n",
    "#         ColSideColors = pbmc3k$analysis$clust5$mapping$Cluster, scale = \"none\")\n",
    "\n",
    "#colnames(toplot.mat)[1:2]\n",
    "#highlightCell<-as.data.frame(NULL)\n",
    "toplot.mat.hcmp<-toplot.mat\n",
    "diag(toplot.mat.hcmp)<- -Inf\n",
    "mark.fun<-which.max\n",
    "#mark.fun2<-function(v){which(v==sort(v,decreasing=TRUE)[2])}\n",
    "highlightCell<-cbind(a=as.data.frame(c(1:NROW(toplot.mat),apply(toplot.mat.hcmp,2,mark.fun))),\n",
    "                     b=c(apply(toplot.mat.hcmp,1,mark.fun),1:NROW(toplot.mat)),\n",
    "                     c=c(rep(\"violet\",NROW(toplot.mat)),rep(\"green1\",NROW(toplot.mat))))\n",
    "#Double marked in third color\n",
    "hcpairs<-paste(highlightCell[,1],highlightCell[,2],sep=\"_\")\n",
    "highlightCell<-rbind(highlightCell,\n",
    "      cbind(highlightCell[duplicated(hcpairs),1:2],c=\"darkgreen\")) \n",
    "\n",
    "cutoff<-0.09\n",
    "\n",
    "f.dev<-list(function()png(paste0(heatmap.filepref,\"_clustering_heatmap_01.png\"),width=2400,height=2400,pointsize=24),\n",
    "            function()psz(12,res=200))\n",
    ".fdn<-0            \n",
    "for (fd in f.dev) { \n",
    ".fdn<-.fdn+1; fd()\n",
    "h3.res<-heatmap3(#t(t(h0.odds.ratio.mat[sub.range,sub.range])/colSums(samp.len.cnts[cnt.range,])), \n",
    "         10^toplot.mat,\n",
    "         #dist.matrix,\n",
    "         Rowv=NULL,symm=TRUE,\n",
    "         col = colorRampPalette(c(\"navy\",\n",
    "                                  \"white\",\n",
    "                                  #\"blue\",\"white\",\"firebrick3\", \n",
    "                                  \"firebrick4\"))(1024),\n",
    "         hclustfun=function(...) fastcluster::hclust(as.dist(dist.matrix),   \n",
    "                                                    #method=\"complete\"\n",
    "                                                    #method=\"single\"\n",
    "                                                    method=\"ward.D2\"\n",
    "                                                    ),\n",
    "         ColSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         RowSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         ColSideLabs = \".strain.\", #\"full.strain\",\n",
    "         RowSideLabs = \"rep.name\",\n",
    "         #RowSideLabs = colnames(h0.odds.ratio.mat[sub.range,sub.range]),\n",
    "         ColAxisColors = 1,#my.colors[as.integer(factor(samp.labs)),2],\n",
    "         RowAxisColors = 1,\n",
    "         #RowSideLabs = samp.labs,\n",
    "         scale=\"none\",\n",
    "         balanceColor = FALSE,\n",
    "         margins=c(12,12),\n",
    "         cexRow=0.6,\n",
    "         cexCol=0.6,\n",
    "         ColSideCut=cutoff,\n",
    "         #ColSideWitdh=1.8,         \n",
    "         #main=\"Clustering according to single-read-averaged odds ratio (log)\",#expression(), #\"[10]*\")\"),\n",
    "         ##main=bquote(paste('Clustering according to single-read-normalized odds ratio (log'['10']*')')),\n",
    "         #legendbar.mar=c(6,1,5,0),\n",
    "         #legendbar.xlab=\"signle.read.normalized\\nodds.ratio\",\n",
    "         main=\"Clustering according to K-L divergence-like distance\",\n",
    "         sub=p(\"(ma.smoothing.window=\",ma.window.size,\", clust.cutoff=\",round(cutoff,3),\")\"),\n",
    "         #legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(exp(-D_KLsym))),\n",
    "         legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(10^(-D_KLsym))),\n",
    "         legendbar.mar=c(6,1,5,0),\n",
    "         #cex=2,\n",
    "         #cex.lab=2,\n",
    "         keep.dendro=TRUE,    \n",
    "         dendr.height.fun=function(h)sqrt(h),\n",
    "         highlightCell=highlightCell,\n",
    "         #revC=FALSE,\n",
    "         na.rm=FALSE\n",
    "         )\n",
    "if (.fdn<2) dev.off()\n",
    "}\n",
    "\n",
    "#   if (TRUE) {\n",
    "#        if (ncol(highlightCell) == 3) {\n",
    "#            highlightCell$lwd <- 1\n",
    "#        }\n",
    "#        highlightCell[, 1] <- h3.res$rowInd[highlightCell[, 1]]\n",
    "#        highlightCell[, 2] <- h3.res$colInd[highlightCell[, 2]]\n",
    "#        rect(highlightCell[, 2] - 0.5, highlightCell[, 1] - 0.5, \n",
    "#            highlightCell[, 2] + 0.5, highlightCell[, 1] + 0.5, \n",
    "#            border = as.character(highlightCell[, 3]), lwd = as.integer(highlightCell[, \n",
    "#                4]))\n",
    "#    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#?options\n",
    "order(h3.res$rowInd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for (i in 1:100)dev.off()\n",
    "printS(\"LCprofiles ordered according to the clustering\")\n",
    "NC<-NCOL(samp.len.cnts.ma); Ncols<-NC%/%5; #print(Ncols)\n",
    "options(repr.plot.width=10,repr.plot.height=10,res=120)\n",
    "..colors<-my.colors[as.integer(factor(samp.color.labs)),2]\n",
    "\n",
    "par(mfrow=c(4,1))\n",
    "for (i in 1:NC){\n",
    " cn<-rev(rank(h3.res$rowInd))[i] \n",
    " plot(samp.len.cnts.ma[1:3000,cn],type=\"l\",pch=\".\",ylab=\"LCp.ma.smooth\") \n",
    " title(colnames(samp.len.cnts.ma)[cn],col.main=..colors[cn]) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf.call(paste0(heatmap.filepref,\"_flowgrams_clustering_order.pdf\"),height=30*NC,width=900)\n",
    "par(mfrow=c(4,1))\n",
    "for (i in 1:NC){\n",
    " cn<-rev(rank(h3.res$rowInd))[i] \n",
    " plot(samp.len.cnts.ma[1:3000,cn],type=\"l\",pch=\".\",ylab=\"LCp.ma.smooth\") \n",
    " title(colnames(samp.len.cnts.ma)[cn],col.main=..colors[cn]) \n",
    "}\n",
    "dev.off()         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Smaller heatmap plot less details\n",
    "printS(\"LCprofiles heatmap (small)\")\n",
    "\n",
    "suppressWarnings(\n",
    "sub.range<-which(grepl(#\"Listeria\",\n",
    "                       #\"Salmonella\",\n",
    "                       \"^rep\",\n",
    "                        #\"_12067|_42222/180|_N53-1|_EGDe|\",#\"_0112F|_4//74|_C5|_LO28|_u292\",\n",
    "                       rep.assign$full.strain)\n",
    "                #&(!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    ")\n",
    "sub.range\n",
    "\n",
    "#\"Salmonella_typhimurium_u292\"\n",
    "#sub.range<-1:288\n",
    "\n",
    "prepare.sub.data.to.heatmap(sub.range)\n",
    "\n",
    "psz(8,res=200)\n",
    "#heatmap3(res.log.scaled.sparse$L.svd$vt[1:8,] * res.log.scaled.sparse$L.svd$d[1:8], \n",
    "#         ColSideColors = pbmc3k$analysis$clust5$mapping$Cluster, scale = \"none\")\n",
    "\n",
    "cutoff<-0.09\n",
    "\n",
    "f.dev<-list(function()png(paste0(heatmap.filepref,\"_clustering_heatmap_02.png\"),width=1600,height=1600,pointsize=36),\n",
    "            function()psz(8,res=200)); \n",
    ".fdn<-0            \n",
    "for (fd in f.dev) { \n",
    ".fdn<-.fdn+1; fd()\n",
    "h3.res<-heatmap3(#t(t(h0.odds.ratio.mat[sub.range,sub.range])/colSums(samp.len.cnts[cnt.range,])), \n",
    "         10^toplot.mat,\n",
    "         #dist.matrix,\n",
    "         Rowv=NULL,symm=TRUE,\n",
    "         col = colorRampPalette(c(\"navy\", \"white\",#\"white\",\"firebrick3\", \n",
    "                                  \"firebrick4\"))(1024),\n",
    "         hclustfun=function(...) fastcluster::hclust(as.dist(dist.matrix),   \n",
    "                                                    #method=\"complete\"\n",
    "                                                    #method=\"single\"\n",
    "                                                    method=\"ward.D2\"\n",
    "                                                    ),\n",
    "         ColSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         RowSideColors = my.colors[as.integer(factor(samp.color.labs)),2],\n",
    "         ColSideLabs = \".strain.\", #\"full.strain\",\n",
    "         RowSideLabs = \"rep.name\",\n",
    "         #RowSideLabs = colnames(h0.odds.ratio.mat[sub.range,sub.range]),\n",
    "         ColAxisColors = 1,#my.colors[as.integer(factor(samp.labs)),2],\n",
    "         RowAxisColors = 1,\n",
    "         #RowSideLabs = samp.labs,\n",
    "         scale=\"none\",\n",
    "         balanceColor = FALSE,\n",
    "         margins=c(12,12),\n",
    "         cexRow=0.6,\n",
    "         cexCol=0.6,\n",
    "         ColSideCut=cutoff,\n",
    "         #RowSideCut=0.2,\n",
    "         #ColSideWitdh=1.8,         \n",
    "         main=\"Clustering according to K-L divergence-like distance\",\n",
    "         sub=p(\"(ma.smoothing.window=\",ma.window.size,\", clust.cutoff=\",round(cutoff,3),\")\"),\n",
    "         #legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(exp(-D_KLsym))),\n",
    "         legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(10^(-D_KLsym))),\n",
    "         legendbar.mar=c(6,1,5,0),         \n",
    "         #cex=2,\n",
    "         #cex.lab=2,\n",
    "         keep.dendro=TRUE,\n",
    "         dendr.height.fun=function(h)sqrt(h),\n",
    "         na.rm=FALSE\n",
    "         )\n",
    "if (.fdn<2) dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NProb\n",
    "$$D_{\\mathrm{nprob}}(LCp_1\\|Dp_2) = -\\sum_i \\frac{LCp_1(i)}{\\#LCp_1} \\, \\log_{10}{Dp_2(i)} \\,=\\, -\\sum_i P_{LCp_1(i)} \\, \\log_{10}{Dp_2(i)}, $$\n",
    "\n",
    "Kullback-Leiber divergence-like distance\n",
    "\n",
    "$$D_{\\mathrm{KL}}(LCp_1\\|Dp_2) = - \\sum_i P_{LCp_1(i)} \\, \\log_{10}\\frac{Dp_2(i)}{Dp_1(i)},$$\n",
    "\n",
    "( which is similar to \n",
    "\n",
    "$$D_{\\mathrm{KL}}(P\\|Q) = \\sum_i P(i) \\, \\log\\frac{P(i)}{Q(i)}.$$\n",
    ")\n",
    "\n",
    "Symmetrised K-L divergence which is symmetric and non-negative.\n",
    "<!-- $$D_{\\mathrm{KL}}(P\\|Q) + D_{\\mathrm{KL}}(Q\\|P)\\, \\!$$ !-->\n",
    "\n",
    "$$D_{\\mathrm{KLsym}}(LCp_1,LCp_2) \\,=\\, \\frac{D_{\\mathrm{KL}}(LCp_1\\|Dp_2) + D_{\\mathrm{KL}}(LCp_2\\|Dp_1)}{2}\\,.$$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "################################################################################\n",
    "################################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stopNbEval(1) ## Saves R image and cleans R environment so the following code is void"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##\n",
    "##  END OF VERIFIED EXECUTABLE CODE\n",
    "##\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n",
    "################################################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO inspect repBC94A, repBC87A, repBC89C why two first are more similar to each other than last two\n",
    "#\n",
    "head(rep.assign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#?heatmap3  \n",
    "\n",
    "toplot.mat.hcmp<-toplot.mat\n",
    "#diag(toplot.mat.hcmp)<- -Inf\n",
    "#highlightCell<-cbind(a=as.data.frame(c(1:NROW(toplot.mat.hcmp),apply(toplot.mat.hcmp,2,which.max))),\n",
    "#                     b=c(apply(toplot.mat.hcmp,1,which.max),1:NROW(toplot.mat.hcmp)),\n",
    "#                     c=c(rep(\"violet\",NROW(toplot.mat)),rep(\"green1\",NROW(toplot.mat))))\n",
    "toplot.df<-as.data.frame(toplot.mat.hcmp)\n",
    "full.strain.colname<-\"full.strain\"\n",
    "#full.strain.colname<-\"strain\"\n",
    "rep.assign.cn<-rep.assign\n",
    "rownames(rep.assign.cn)<-rep.assign.cn$rep.name\n",
    "correctness<-as.data.frame(t(mapply(function(col,col.full.strain){\n",
    "        fs.sorted<-rep.assign.cn[colnames(toplot.df)[order(col,decreasing=TRUE)],full.strain.colname]\n",
    "        cluster.size<-sum(rep.assign.cn[colnames(toplot.df),full.strain.colname]==col.full.strain)\n",
    "        c(corr=mean(fs.sorted[1:cluster.size]==col.full.strain),size=cluster.size)\n",
    "        },toplot.df,rep.assign.cn[colnames(toplot.df),full.strain.colname],\n",
    "      SIMPLIFY=TRUE)))\n",
    "#Double marked in third color\n",
    "#hcpairs<-paste(highlightCell[,1],highlightCell[,2],sep=\"_\")\n",
    "#highlightCell<-rbind(highlightCell,\n",
    "#      cbind(highlightCell[duplicated(hcpairs),1:2],c=\"darkgreen\")) \n",
    "#summary(correctness)\n",
    "\n",
    "#correctness\n",
    "#table(correctness$corr)\n",
    "with(correctness,\n",
    "           c(S1=sum(corr*size)/sum(size),\n",
    "             S2=sum((1-corr)*size)/sum((NROW(correctness)-size)))\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#toplot.mat[\"\"]\n",
    "#h0.odds.ratio.mat<- as.matrix(t(res[-(1:2),])-as.vector(as.matrix(res[1,]))) #)/2  #h0.sample\n",
    "\n",
    "#\"repBC28C\" - N53-1\n",
    "#\"repBC29B\" - 42222/180\n",
    "\n",
    "10^h0.odds.ratio.mat[\"repBC28C\",\"repBC29B\"]\n",
    "10^h0.odds.ratio.mat[\"repBC29B\",\"repBC28C\"]\n",
    "\n",
    "#Things are more similar to repBC28C/N53-1 than the other way round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(as.vector(t(res[1,])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO inspect repBC04A repBC08A  repBC28C repBC20C  repBC28A  repBC05A repBC87A\n",
    "#             repBC12A repBC17A\n",
    "toinsp.names<-c(\"repBC04A\",\"repBC08A\",\"repBC28C\",\"repBC20C\",\"repBC28A\",\"repBC05A\",\"repBC87A\")\n",
    "res[1,toinsp.names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names(h3.res)\n",
    "class(h3.res$Colv)\n",
    "head(cutree(as.hclust(h3.res$Colv),h=0.2)[h3.res$colInd])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.range<-which(#grepl(#\"Salmonella\",\n",
    "                 #      \"Bacillus|Salmonella|Listeria\",\n",
    "                 #      rep.assign$full.strain)\n",
    "                 #&(!grepl(#\"Salmonella\",\n",
    "                 #      \"subtilis|lichniformis\",\n",
    "                 #      rep.assign$full.strain))\n",
    "                 #&\n",
    "                (!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    "mean(colSums(samp.len.cnts[,sub.range]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "h3.res.all.params.list<-list(\n",
    "\"h3.res.all.frac0.5\"=c(0.5,TRUE),\n",
    "\"h3.res.all.frac0.1\"=c(0.1,TRUE),\n",
    "\"h3.res.all.frac0.1s2\"=c(0.1,TRUE),\n",
    "\"h3.res.all.frac0.02\"=c(0.02,TRUE),\n",
    "\"h3.res.all.frac0.01\"=c(0.01,TRUE),\n",
    "\"h3.res.all\"=c(1,FALSE),\n",
    "\"h3.res.all.noC\"=c(1,TRUE)\n",
    ")\n",
    "\n",
    "#h3.res.all.params.res.compl<-h3.res.all.params.list\n",
    "h3.res.all.params.res<-h3.res.all.params.list\n",
    "\n",
    "for (h3.params.name in names(h3.res.all.params.list)){\n",
    "\n",
    "suppressWarnings(\n",
    "sub.range<-which(#grepl(#\"Salmonella\",\n",
    "                 #      \"Bacillus|Salmonella|Listeria\",\n",
    "                 #      rep.assign$full.strain)\n",
    "                 #&(!grepl(#\"Salmonella\",\n",
    "                 #      \"subtilis|lichniformis\",\n",
    "                 #      rep.assign$full.strain))\n",
    "                 #&\n",
    "                (!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    ")\n",
    "if (!h3.res.all.params.list[[h3.params.name]][2]){\n",
    "    sub.range<-1:NCOL(samp.len.cnts)\n",
    "}\n",
    "print(length(sub.range)/NCOL(samp.len.cnts))\n",
    "\n",
    "#frac<-0.1\n",
    "#frac<-1\n",
    "frac<-h3.res.all.params.list[[h3.params.name]][1]\n",
    "print(p(\"frac=\",frac))\n",
    "    \n",
    "if (TRUE) {#frac!=1.0) {\n",
    "\n",
    "    ##samp.len.cnts.org<-samp.len.cnts\n",
    "    ##Subsample \n",
    "    l<-1:NROW(samp.len.cnts.org)\n",
    "    samp.len.cnts<-sapply(1:NCOL(samp.len.cnts.org),function(i){\n",
    "        src<-rep(l,samp.len.cnts.org[,i])\n",
    "        tabulate(sample(src,length(src)*frac),NROW(samp.len.cnts.org))\n",
    "    })\n",
    "    dim(samp.len.cnts)\n",
    "    if (frac==1.) samp.len.cnts<-samp.len.cnts.org\n",
    "\n",
    "    print(summary(colSums(samp.len.cnts)/colSums(samp.len.cnts.org)))\n",
    "    print(summary(colSums(samp.len.cnts)))\n",
    "\n",
    "    #samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)as.integer(ma(r,20)))    \n",
    "    samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)ma(r,20)) #30\n",
    "\n",
    "    ##\n",
    "    ## Normalize database of known profiles\n",
    "    ##\n",
    "\n",
    "    samp.len.cnts.ma.norm<-apply(samp.len.cnts.ma,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "    samp.len.cnts.ma.norm[is.na(samp.len.cnts.ma.norm)]<-0\n",
    "\n",
    "    samp.len.cnts.norm<-apply(samp.len.cnts,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "    samp.len.cnts.norm[is.na(samp.len.cnts.norm)]<-0                             \n",
    "\n",
    "    all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-11)\n",
    "    all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-7)        \n",
    "                              \n",
    "    h0.odds.ratio.mat<-prepare.h0.odds.ratio.mat(150:3000,0.01)                              \n",
    "}    \n",
    "\n",
    "prepare.sub.data.to.heatmap(sub.range)\n",
    "\n",
    "psz(26,res=150)\n",
    "#heatmap3(res.log.scaled.sparse$L.svd$vt[1:8,] * res.log.scaled.sparse$L.svd$d[1:8], \n",
    "#         ColSideColors = pbmc3k$analysis$clust5$mapping$Cluster, scale = \"none\")\n",
    "\n",
    "cutoff<-0.09\n",
    "#h3.res.all.frac0.1s2<-\n",
    "#h3.res.all.params.res.compl[[h3.params.name]]<-\n",
    "h3.res.all.params.res[[h3.params.name]]<-        heatmap3(#t(t(h0.odds.ratio.mat[sub.range,sub.range])/colSums(samp.len.cnts[cnt.range,])), \n",
    "         10^toplot.mat,\n",
    "         #dist.matrix,\n",
    "         Rowv=NULL,symm=TRUE,\n",
    "         col = colorRampPalette(c(\"navy\",\"blue\", \"white\", \n",
    "                                  \"firebrick3\", \"firebrick4\"))(1024),\n",
    "         hclustfun=function(...) fastcluster::hclust(as.dist(dist.matrix),   \n",
    "                                                    #method=\"complete\"\n",
    "                                                    #method=\"single\"\n",
    "                                                    method=\"ward.D2\"\n",
    "                                                    ),\n",
    "         ColSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         RowSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         ColSideLabs = \".strain.\", #\"full.strain\",\n",
    "         RowSideLabs = \"rep.name\",\n",
    "         #RowSideLabs = colnames(h0.odds.ratio.mat[sub.range,sub.range]),\n",
    "         ColAxisColors = 1,#my.colors[as.integer(factor(samp.labs)),2],\n",
    "         RowAxisColors = 1,\n",
    "         #RowSideLabs = samp.labs,\n",
    "         scale=\"none\",\n",
    "         balanceColor = FALSE,\n",
    "         margins=c(12,12),\n",
    "         cexRow=0.4,\n",
    "         cexCol=0.4,\n",
    "         ColSideCut=cutoff,        \n",
    "         ##ColSideWitdh=1.8,         \n",
    "         #main=\"Clustering according to single-read-normalized odds ratio\",\n",
    "         #legendbar.mar=c(6,1,5,0),\n",
    "         #legendbar.xlab=\"signle.read.normalized.odds.ratio\",\n",
    "         main=p(\"Clustering according to K-L divergence-like distance\", ifelse(h3.params.name!=\"h3.res.all.noC\",p(\" - \", h3.params.name),\"\")),\n",
    "         sub=p(\"(ma.smoothing.window=\",ma.window.size,\", clust.cutoff=\",round(cutoff,3),\")\"),\n",
    "         legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(exp(-D_KLsym))),\n",
    "         legendbar.mar=c(6,1,5,0),\n",
    "         keep.dendro=TRUE,    \n",
    "         #dendr.height.fun=function(h)sqrt(h),\n",
    "         #cex=2,\n",
    "         #cex.lab=2,\n",
    "        \n",
    "         na.rm=FALSE\n",
    "         )                              \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "png(\"/dev/null\")\n",
    "\n",
    "eff.frac<-0.2\n",
    "#eff.frac<-0.1\n",
    "h3.res.all.eff.params.list<-list(\n",
    "\"h3.res.all.frac0.2\"=c(eff.frac,TRUE),\n",
    "\"h3.res.all.frac0.2s1\"=c(eff.frac,TRUE),\n",
    "\"h3.res.all.frac0.2s2\"=c(eff.frac,TRUE),\n",
    "\"h3.res.all.frac0.2s3\"=c(eff.frac,TRUE),\n",
    "\"h3.res.all.frac0.2s4\"=c(eff.frac,TRUE),\n",
    "\"h3.res.all\"=c(1,FALSE),\n",
    "\"h3.res.all.noC\"=c(1,TRUE)\n",
    ")\n",
    "\n",
    "#h3.res.all.params.res.compl<-h3.res.all.params.list\n",
    "h3.res.all.eff.params.res<-h3.res.all.params.list\n",
    "\n",
    "for (h3.params.name in names(h3.res.all.eff.params.list)){\n",
    "\n",
    "suppressWarnings(\n",
    "sub.range<-which(#grepl(#\"Salmonella\",\n",
    "                 #      \"Bacillus|Salmonella|Listeria\",\n",
    "                 #      rep.assign$full.strain)\n",
    "                 #&(!grepl(#\"Salmonella\",\n",
    "                 #      \"subtilis|lichniformis\",\n",
    "                 #      rep.assign$full.strain))\n",
    "                 #&\n",
    "                (!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    ")\n",
    "if (!h3.res.all.eff.params.list[[h3.params.name]][2]){\n",
    "    sub.range<-1:NCOL(samp.len.cnts)\n",
    "}\n",
    "print(length(sub.range)/NCOL(samp.len.cnts))\n",
    "\n",
    "#frac<-0.1\n",
    "#frac<-1\n",
    "frac<-h3.res.all.eff.params.list[[h3.params.name]][1]\n",
    "print(p(\"frac=\",frac))\n",
    "    \n",
    "if (TRUE) {#frac!=1.0) {\n",
    "\n",
    "    ##samp.len.cnts.org<-samp.len.cnts\n",
    "    ##Subsample \n",
    "    l<-1:NROW(samp.len.cnts.org)\n",
    "    samp.len.cnts<-sapply(1:NCOL(samp.len.cnts.org),function(i){\n",
    "        src<-rep(l,samp.len.cnts.org[,i])\n",
    "        tabulate(sample(src,length(src)*frac),NROW(samp.len.cnts.org))\n",
    "    })\n",
    "    dim(samp.len.cnts)\n",
    "    if (frac==1.) samp.len.cnts<-samp.len.cnts.org\n",
    "\n",
    "    print(summary(colSums(samp.len.cnts)/colSums(samp.len.cnts.org)))\n",
    "    print(summary(colSums(samp.len.cnts)))\n",
    "\n",
    "    #samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)as.integer(ma(r,20)))    \n",
    "    samp.len.cnts.ma<-apply(samp.len.cnts,2,function(r)ma(r,20)) #30\n",
    "\n",
    "    ##\n",
    "    ## Normalize database of known profiles\n",
    "    ##\n",
    "\n",
    "    samp.len.cnts.ma.norm<-apply(samp.len.cnts.ma,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "    samp.len.cnts.ma.norm[is.na(samp.len.cnts.ma.norm)]<-0\n",
    "\n",
    "    samp.len.cnts.norm<-apply(samp.len.cnts,2,function(v)as.numeric(v)/sum(v,na.rm=TRUE))\n",
    "    samp.len.cnts.norm[is.na(samp.len.cnts.norm)]<-0                             \n",
    "\n",
    "    all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-11)\n",
    "    all(abs(colSums(samp.len.cnts.ma.norm)-1)<10^-7)        \n",
    "                              \n",
    "    h0.odds.ratio.mat<-prepare.h0.odds.ratio.mat(150:3000,0.01)                              \n",
    "}    \n",
    "\n",
    "prepare.sub.data.to.heatmap(sub.range)\n",
    "\n",
    "psz(26,res=150)\n",
    "#heatmap3(res.log.scaled.sparse$L.svd$vt[1:8,] * res.log.scaled.sparse$L.svd$d[1:8], \n",
    "#         ColSideColors = pbmc3k$analysis$clust5$mapping$Cluster, scale = \"none\")\n",
    "\n",
    "cutoff<-0.09\n",
    "#h3.res.all.frac0.1s2<-\n",
    "#h3.res.all.params.res.compl[[h3.params.name]]<-\n",
    "h3.res.all.eff.params.res[[h3.params.name]]<-        heatmap3(#t(t(h0.odds.ratio.mat[sub.range,sub.range])/colSums(samp.len.cnts[cnt.range,])), \n",
    "         10^toplot.mat,\n",
    "         #dist.matrix,\n",
    "         Rowv=NULL,symm=TRUE,\n",
    "         col = colorRampPalette(c(\"navy\",\"blue\", \"white\", \n",
    "                                  \"firebrick3\", \"firebrick4\"))(1024),\n",
    "         hclustfun=function(...) fastcluster::hclust(as.dist(dist.matrix),   \n",
    "                                                    #method=\"complete\"\n",
    "                                                    #method=\"single\"\n",
    "                                                    method=\"ward.D2\"\n",
    "                                                    ),\n",
    "         ColSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         RowSideColors = my.colors[as.integer(factor(samp.color.labs)),2],         \n",
    "         ColSideLabs = \".strain.\", #\"full.strain\",\n",
    "         RowSideLabs = \"rep.name\",\n",
    "         #RowSideLabs = colnames(h0.odds.ratio.mat[sub.range,sub.range]),\n",
    "         ColAxisColors = 1,#my.colors[as.integer(factor(samp.labs)),2],\n",
    "         RowAxisColors = 1,\n",
    "         #RowSideLabs = samp.labs,\n",
    "         scale=\"none\",\n",
    "         balanceColor = FALSE,\n",
    "         margins=c(12,12),\n",
    "         cexRow=0.4,\n",
    "         cexCol=0.4,\n",
    "         ColSideCut=cutoff,        \n",
    "         ##ColSideWitdh=1.8,         \n",
    "         #main=\"Clustering according to single-read-normalized odds ratio\",\n",
    "         #legendbar.mar=c(6,1,5,0),\n",
    "         #legendbar.xlab=\"signle.read.normalized.odds.ratio\",\n",
    "         main=p(\"Clustering according to K-L divergence-like distance\", ifelse(h3.params.name!=\"h3.res.all.noC\",p(\" - \", h3.params.name),\"\")),\n",
    "         sub=p(\"(ma.smoothing.window=\",ma.window.size,\", clust.cutoff=\",round(cutoff,3),\")\"),\n",
    "         legendbar.xlab=paste(\"read.averaged.odds.ratio\\n\",expression(exp(-D_KLsym))),\n",
    "         legendbar.mar=c(6,1,5,0),\n",
    "         keep.dendro=TRUE,    \n",
    "         #dendr.height.fun=function(h)sqrt(h),\n",
    "         #cex=2,\n",
    "         #cex.lab=2,\n",
    "        \n",
    "         na.rm=FALSE\n",
    "         )                              \n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.range<-which(#grepl(#\"Salmonella\",\n",
    "                 #      \"Bacillus|Salmonella|Listeria\",\n",
    "                 #      rep.assign$full.strain)\n",
    "                 #&(!grepl(#\"Salmonella\",\n",
    "                 #      \"subtilis|lichniformis\",\n",
    "                 #      rep.assign$full.strain))\n",
    "                 #&\n",
    "                (!grepl(\"C$\",as.character(rep.assign$rep.name)))\n",
    "                )\n",
    "length(sub.range)\n",
    "dim(samp.len.cnts)\n",
    "print(summary(colSums(samp.len.cnts[,sub.range])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(ct.res)\n",
    "length(sub.range)\n",
    "head(cbind(names(ct.res),as.character(rep.assign$.strain.[sub.range][h3.res.all.noC$colInd])))\n",
    "all((names(ct.res)==as.character(rep.assign$.strain.[sub.range][h3.res.all.noC$colInd])))\n",
    "#sort(rep.assign$.strain.[sub.range])[sort(names(ct.res))!=sort(rep.assign$.strain.[sub.range])]\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "names(h3.res.all.noC)\n",
    "h3.res.all.noC$cutTable\n",
    "rowInd\n",
    "#mean(h3.res.all.noC$rowInd==h3.res.all.noC$colInd)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "similar.labels\n",
    "are.similar.strains.sim<-function(n1,n2,simlabels=similar.labels){\n",
    "    stopifnot(length(n1)<=1)    \n",
    "    for (v in simlabels) {\n",
    "      to.ren<-n1%in%v\n",
    "      n1[to.ren]<-v[1]\n",
    "      to.ren<-n2%in%v\n",
    "      n2[to.ren]<-v[1]\n",
    "    }\n",
    "    n1==n2\n",
    "}\n",
    "are.similar.strains.def<-function(n1,n2,simlabels=similar.labels){n1==n2}\n",
    "\n",
    "#are.similar.strains<-are.similar.strains.sim\n",
    "are.similar.strains<-are.similar.strains.def"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list2env(h3.res.all.params.res,as.environment(-1))\n",
    "\n",
    "clustering.perf.problematic<-function(h3.res.all,cut=0.09){\n",
    "  cind<-h3.res.all$colInd\n",
    "  ct.res<-cutree(as.hclust(h3.res.all$Colv),h=sqrt(cut))[h3.res.all$colInd]\n",
    "  s<-0;s2<-0\n",
    "  v<-lapply(1:length(ct.res),function(i){\n",
    "    #ids<-which(names(ct.res)[i]==names(ct.res))\n",
    "    ids<-which(are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    s<<-s+length(ids)-1\n",
    "    pr.ids<-ids[ct.res[i]!=ct.res[ids]]\n",
    "    if(length(pr.ids)>0)\n",
    "        cbind(cind[i],cind[pr.ids])\n",
    "    else NULL    \n",
    "    #ct.res[i]==ct.res[ids2]        \n",
    "    })\n",
    "    \n",
    "  v2<-lapply(1:length(ct.res),function(i){\n",
    "    #ids2<-which(names(ct.res)[i]!=names(ct.res))\n",
    "    ids2<-which(!are.similar.strains(names(ct.res)[i],names(ct.res)))    \n",
    "    s2<<-s2+length(ids2)\n",
    "    pr.ids2<-ids2[ct.res[i]==ct.res[ids2]]\n",
    "    if(length(pr.ids2)>0)\n",
    "        cbind(cind[i],cind[pr.ids2])\n",
    "    else NULL    \n",
    "    #ct.res[i]==ct.res[ids2]        \n",
    "    })\n",
    "    \n",
    "print(p(\"s=\",s,\" s2=\",s2))\n",
    "list(do.call(rbind,v[!sapply(v,is.null)]),do.call(rbind,v2[!sapply(v2,is.null)]),ct.res)\n",
    "}\n",
    "\n",
    "are.similar.strains<-are.similar.strains.sim\n",
    "#are.similar.strains<-are.similar.strains.def\n",
    "\n",
    "cl.perf.pr.all<-clustering.perf.problematic(h3.res.all.noC,0.09)\n",
    "ct.res<-cl.perf.pr.all[[3]]\n",
    "cl.perf.pr<-cl.perf.pr.all[[1]]\n",
    "#print(cl.perf.pr)\n",
    "#cl.perf.pr<-cl.perf.pr.all[[2]]\n",
    "#print(cl.perf.pr)\n",
    "if(!is.null(cl.perf.pr)){\n",
    "prob.reps<-as.matrix(cl.perf.pr)\n",
    "\n",
    "#names()\n",
    "prob.reps[]<-as.character(rep.assign$rep.name[sub.range][cl.perf.pr])\n",
    "length(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))\n",
    "t(t(table(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))))\n",
    "} else print(NULL)\n",
    "#dim(prob.reps)\n",
    "#cbind(prob.reps,as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]),\n",
    "#      as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,2]]))\n",
    "\n",
    "cl.perf.pr<-cl.perf.pr.all[[2]]\n",
    "if(!is.null(cl.perf.pr)){\n",
    "prob.reps<-as.matrix(cl.perf.pr)\n",
    "prob.reps[]<-as.character(rep.assign$rep.name[sub.range][cl.perf.pr])\n",
    "length(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))\n",
    "#t(t(table(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))))\n",
    "pr.df<-cbind(prob.reps,as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]),\n",
    "      as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,2]]))\n",
    "#t(apply(pr.df[,3:4],1,sort))\n",
    "t(t(table(apply(t(apply(pr.df[,3:4],1,sort)),1,function(...)paste(...,collapse=\" -- \")))))\n",
    "                } else print(NULL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cl.perf.pr<-cl.perf.pr.all[[1]]\n",
    "prob.reps<-as.matrix(cl.perf.pr)\n",
    "#prob.reps\n",
    "#names()\n",
    "prob.reps[]<-as.character(rep.assign$rep.name[sub.range][cl.perf.pr])\n",
    "length(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))\n",
    "t(t(table(as.character(rep.assign$.strain.[sub.range][cl.perf.pr[,1]]))))\n",
    "#v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(names(ct.res)[90]==names(ct.res))\n",
    "sum(are.similar.strains(names(ct.res)[90],names(ct.res)))\n",
    "names(ct.res)[90]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(ct.res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(do.call(c,lapply(1:length(ct.res),function(i)which(!are.similar.strains.def(names(ct.res)[i],names(ct.res))))))\n",
    "                        \n",
    "v<-sapply(1:length(ct.res),function(i){\n",
    "    ids<-which(are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    ids2<-which(!are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    c(sum(ct.res[i]==ct.res[ids]),\n",
    "      length(ids),\n",
    "      sum(ct.res[i]!=ct.res[ids2]),\n",
    "      length(ids2))\n",
    "    })\n",
    "sum(v[2,]-1)                        \n",
    "                        \n",
    "clustering.perf<-function(h3.res.all,cuts=seq(0,1,by=0.01)){\n",
    "res<-lapply(cuts,function(ctval){\n",
    "#res<-lapply(cuts,function(ctval){    \n",
    "ct.res<-cutree(as.hclust(h3.res.all$Colv),h=sqrt(ctval))[h3.res.all$colInd]\n",
    "v<-sapply(1:length(ct.res),function(i){\n",
    "    ids<-which(are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    ids2<-which(!are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    c(sum(ct.res[i]==ct.res[ids]),\n",
    "      length(ids),\n",
    "      sum(ct.res[i]!=ct.res[ids2]),\n",
    "      length(ids2))\n",
    "    })\n",
    "#print(sum(v[2,]-1))\n",
    "#print(sum(v[4,]))\n",
    "r<-c(sum(v[1,]-1)/sum(v[2,]-1),sum(v[3,])/sum(v[4,]))\n",
    "#1-sqrt(1-r)    \n",
    "list(a=r,b=sum(v[2,]-1),c=sum(v[4,]))\n",
    "#c(-sum(v[2,]-1)+sum(v[1,]-1),-sum(v[4,])+sum(v[3,]))\n",
    "})                        \n",
    "print(paste(\"same=\",unique(sapply(res,function(r)r[[2]]))))\n",
    "print(paste(\"diff=\",unique(sapply(res,function(r)r[[3]]))))\n",
    "res<-lapply(res,function(r)r[[1]])\n",
    "do.call(cbind,res)\n",
    "}\n",
    "    \n",
    "clustering.perf(h3.res.all.params.res[[\"h3.res.all.noC\"]],cuts)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dev.off()\n",
    "list2env(h3.res.all.params.res,as.environment(-1))\n",
    "#list2env(h3.res.all.params.res.compl,as.environment(-1))\n",
    "\n",
    "#names(h3.res.all)\n",
    "#class(h3.res.all$Colv)\n",
    "clustering.perf<-function(h3.res.all,cuts=seq(0,1,by=0.01)){\n",
    "res<-mclapply(cuts,function(ctval){\n",
    "#res<-lapply(cuts,function(ctval){    \n",
    "ct.res<-cutree(as.hclust(h3.res.all$Colv),h=sqrt(ctval))[h3.res.all$colInd]\n",
    "v<-sapply(1:length(ct.res),function(i){\n",
    "    ids<-which(are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    ids2<-which(!are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    c(sum(ct.res[i]==ct.res[ids]),\n",
    "      length(ids),\n",
    "      sum(ct.res[i]!=ct.res[ids2]),\n",
    "      length(ids2))\n",
    "    })\n",
    "#print(sum(v[2,]-1))\n",
    "#print(sum(v[4,]))\n",
    "r<-c(sum(v[1,]-1)/sum(v[2,]-1),sum(v[3,])/sum(v[4,]))\n",
    "#1-sqrt(1-r)    \n",
    "list(a=r,b=sum(v[2,]-1),c=sum(v[4,]))\n",
    "#c(-sum(v[2,]-1)+sum(v[1,]-1),-sum(v[4,])+sum(v[3,]))\n",
    "},mc.cores=4)\n",
    "print(paste(\"same=\",unique(sapply(res,function(r)r[[2]]))))\n",
    "print(paste(\"diff=\",unique(sapply(res,function(r)r[[3]]))))\n",
    "res<-lapply(res,function(r)r[[1]])\n",
    "do.call(cbind,res)\n",
    "}\n",
    "\n",
    "icnt<-1\n",
    "for (are.similar.strains in c(are.similar.strains.sim, are.similar.strains.def)){\n",
    "#cuts<-sort(c(seq(0.03,0.3,by=0.01))) #,0.088))\n",
    "cuts<-sort(c(seq(0.0,1,by=0.002))) #,0.088))\n",
    "#cpf<-clustering.perf(h3.res.all,cuts)\n",
    "#cpfnc<-clustering.perf(h3.res.all.noC,cuts)\n",
    "cpff0.5<-clustering.perf(h3.res.all.frac0.5,cuts)\n",
    "cpff0.1<-clustering.perf(h3.res.all.frac0.1,cuts)\n",
    "cpff0.1s2<-clustering.perf(h3.res.all.frac0.1s2,cuts)\n",
    "cpff0.02<-clustering.perf(h3.res.all.frac0.02,cuts)\n",
    "cpff0.01<-clustering.perf(h3.res.all.frac0.01,cuts)\n",
    "\n",
    "#cuts<-sort(c(seq(0.05,0.35,by=0.005))) #,0.088))\n",
    "cpf<-clustering.perf(h3.res.all,cuts)\n",
    "cpfnc<-clustering.perf(h3.res.all.noC,cuts)\n",
    "\n",
    "if (icnt==1)    \n",
    "cpf.similar.strains.sim.list<-list(\n",
    "    cpff0.5=cpff0.5,cpff0.1=cpff0.1,cpff0.1s2=cpff0.1s2,cpff0.02=cpff0.02,cpff0.01=cpff0.01,\n",
    "    cpf=cpf,cpfnc=cpfnc\n",
    ")\n",
    "else\n",
    "cpf.similar.strains.def.list<-list(\n",
    "    cpff0.5=cpff0.5,cpff0.1=cpff0.1,cpff0.1s2=cpff0.1s2,cpff0.02=cpff0.02,cpff0.01=cpff0.01,\n",
    "    cpf=cpf,cpfnc=cpfnc\n",
    ")\n",
    "icnt<-icnt+1   \n",
    "print(\"--------\")\n",
    "}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dev.off()\n",
    "list2env(h3.res.all.eff.params.res,as.environment(-1))\n",
    "#list2env(h3.res.all.params.res.compl,as.environment(-1))\n",
    "\n",
    "clustering.perf<-function(h3.res.all,cuts=seq(0,1,by=0.01)){\n",
    "res<-mclapply(cuts,function(ctval){\n",
    "#res<-lapply(cuts,function(ctval){    \n",
    "ct.res<-cutree(as.hclust(h3.res.all$Colv),h=sqrt(ctval))[h3.res.all$colInd]\n",
    "v<-sapply(1:length(ct.res),function(i){\n",
    "    ids<-which(are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    ids2<-which(!are.similar.strains(names(ct.res)[i],names(ct.res)))\n",
    "    c(sum(ct.res[i]==ct.res[ids]),\n",
    "      length(ids),\n",
    "      sum(ct.res[i]!=ct.res[ids2]),\n",
    "      length(ids2))\n",
    "    })\n",
    "#print(sum(v[2,]-1))\n",
    "#print(sum(v[4,]))\n",
    "r<-c(sum(v[1,]-1)/sum(v[2,]-1),sum(v[3,])/sum(v[4,]))\n",
    "#1-sqrt(1-r)    \n",
    "list(a=r,b=sum(v[2,]-1),c=sum(v[4,]))\n",
    "#c(-sum(v[2,]-1)+sum(v[1,]-1),-sum(v[4,])+sum(v[3,]))\n",
    "},mc.cores=4)\n",
    "print(paste(\"same=\",unique(sapply(res,function(r)r[[2]]))))\n",
    "print(paste(\"diff=\",unique(sapply(res,function(r)r[[3]]))))\n",
    "res<-lapply(res,function(r)r[[1]])\n",
    "do.call(cbind,res)\n",
    "}\n",
    "\n",
    "icnt<-1\n",
    "for (are.similar.strains in c(are.similar.strains.sim, are.similar.strains.def)){\n",
    "#cuts<-sort(c(seq(0.03,0.3,by=0.01))) #,0.088))\n",
    "cuts<-sort(c(seq(0.0,1,by=0.002))) #,0.088))\n",
    "#cpf<-clustering.perf(h3.res.all,cuts)\n",
    "#cpfnc<-clustering.perf(h3.res.all.noC,cuts)\n",
    "cpff0.2<-clustering.perf(h3.res.all.frac0.2,cuts)\n",
    "cpff0.2s1<-clustering.perf(h3.res.all.frac0.2s1,cuts)\n",
    "cpff0.2s2<-clustering.perf(h3.res.all.frac0.2s2,cuts)\n",
    "cpff0.2s3<-clustering.perf(h3.res.all.frac0.2s3,cuts)\n",
    "cpff0.2s4<-clustering.perf(h3.res.all.frac0.2s4,cuts)\n",
    "\n",
    "#cuts<-sort(c(seq(0.05,0.35,by=0.005))) #,0.088))\n",
    "cpf<-clustering.perf(h3.res.all,cuts)\n",
    "cpfnc<-clustering.perf(h3.res.all.noC,cuts)\n",
    "\n",
    "if (icnt==1)    \n",
    "cpf.similar.strains.eff.sim.list<-list(\n",
    "    cpff0.2=cpff0.2,cpff0.2s1=cpff0.2s1,cpff0.2s2=cpff0.2s2,cpff0.2s3=cpff0.2s3,cpff0.2s4=cpff0.2s4,\n",
    "    cpf=cpf,cpfnc=cpfnc\n",
    ")\n",
    "else\n",
    "cpf.similar.strains.eff.def.list<-list(\n",
    "    cpff0.2=cpff0.2,cpff0.2s1=cpff0.2s1,cpff0.2s2=cpff0.2s2,cpff0.2s3=cpff0.2s3,cpff0.2s4=cpff0.2s4,\n",
    "    cpf=cpf,cpfnc=cpfnc\n",
    ")\n",
    "icnt<-icnt+1   \n",
    "print(\"--------\")\n",
    "}    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#psz\n",
    "#lsos()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(12,6,res=300)\n",
    "par(mfcol=c(1,2))\n",
    "list2env(cpf.similar.strains.def.list,as.environment(-1))\n",
    "par(lab = c(10, 10, 7))\n",
    "plot(cpf[2,],cpf[1,],ylab=\"\",xlab=\"\",\n",
    "     type=\"b\",pch=1,lty=3,cex=0.8,col=\"blue\",\n",
    "    xlim=c(0.99,1),ylim=c(0.55,1))\n",
    "#axis(2,at=seq(0,1,by=0.05))\n",
    "lines(cpfnc[2,],cpfnc[1,],type=\"b\",pch=3,lty=1)\n",
    "lines(cpff0.5[2,],cpff0.5[1,],type=\"b\",pch=4,lty=3,col=\"darkgreen\")\n",
    "lines(cpff0.1[2,],cpff0.1[1,],type=\"b\",pch=5,lty=3,col=\"magenta\")\n",
    "lines(cpff0.1s2[2,],cpff0.1s2[1,],type=\"b\",pch=5,lty=3,col=\"purple\")\n",
    "lines(cpff0.02[2,],cpff0.02[1,],type=\"b\",pch=6,lty=3,col=\"violet\")\n",
    "grid()\n",
    "i<-which(cuts%in%c(0.09,0.04,0.014))\n",
    "text(cpfnc[2,i],cpfnc[1,i],p(\"c=\",cuts[i]),col=\"red\",font=2,adj=c(0.5,0))\n",
    "i<-which(cuts%in%c(0.09))\n",
    "title(main=\"a)                        ROC curves for clustering cutoff                          \",line=0.5)\n",
    "title(ylab=\"sensitivity (true pairwise strain eq.)\",line=2.2); \n",
    "title(xlab=\"specificity (true pairwise strain ineq.)\",line=2.2)\n",
    "legend(\"bottomleft\",ncol=2,\n",
    "       c(\"all.rep (avg# 8652)\",  \"wo.rep*C (avg# 8435)\",\n",
    "         \"50%.wo.rep*C(avg# 4326)\",   \"10%.wo.rep*C(avg# 865)\",     \"10%.wo.rep*C(avg# 865)\",\n",
    "         \"2%.wo.rep*C(avg# 173)\"\n",
    "                     #,\"\" #p(p(\"c=\",cuts[i]),\": spec.=\",round(cpfnc[2,i],3),\", sens.=\",round(cpfnc[1,i],3))\n",
    "                     ),lty=c(3,1,3,3,3,3),#,NA),\n",
    "       #cex=c(1,1,1,1,1,0.7),\n",
    "       col=c(\"blue\",\"black\",\"darkgreen\",\"magenta\",\"purple\",\"violet\"),#,\"black\"),\n",
    "       pch=c(1,3,4,5,5,6),#,NA),\n",
    "       cex=0.8, text.width=0.0032)\n",
    "cpfnc.def<-cpfnc\n",
    "list2env(cpf.similar.strains.sim.list,as.environment(-1))\n",
    "par(lab = c(10, 10, 7))\n",
    "plot(cpf[2,],cpf[1,],ylab=\"\",xlab=\"\",type=\"b\",pch=1,lty=3,cex=0.8,col=\"blue\",\n",
    "    xlim=c(0.99,1),ylim=c(0.55,1))\n",
    "#axis(2,at=seq(0,1,by=0.05))\n",
    "lines(cpfnc[2,],cpfnc[1,],type=\"b\",pch=3,lty=1)\n",
    "lines(cpff0.5[2,],cpff0.5[1,],type=\"b\",pch=4,lty=3,col=\"darkgreen\")\n",
    "lines(cpff0.1[2,],cpff0.1[1,],type=\"b\",pch=5,lty=3,col=\"magenta\")\n",
    "lines(cpff0.1s2[2,],cpff0.1s2[1,],type=\"b\",pch=5,lty=3,col=\"purple\")\n",
    "lines(cpff0.02[2,],cpff0.02[1,],type=\"b\",pch=6,lty=3,col=\"violet\")\n",
    "grid()\n",
    "i<-which(cuts%in%c(0.09,0.04,0.014))\n",
    "text(cpfnc[2,i],cpfnc[1,i],p(\"c=\",cuts[i]),col=\"red\",font=2,adj=c(0.5,0))\n",
    "i<-which(cuts%in%c(0.09))\n",
    "title(main=\"b)                        ROC curves for clustering cutoff                          \",line=0.5)\n",
    "#title(ylab=\"sensitivity\",line=2.2); title(xlab=\"specificity\",line=2.2)\n",
    "title(ylab=\"sensitivity (true pairwise strain eq.)\",line=2.2); \n",
    "title(xlab=\"specificity (true pairwise strain ineq.)\",line=2.2)\n",
    "legend(\"bottomleft\",ncol=2,\n",
    "#       c(\"all.rep\",\"wo.rep*C\",\"50%.data.wo.rep*C\",\"10%.data.wo.rep*C\",\"10%.data.wo.rep*C\",\n",
    "#         \"2%.data.wo.rep*C\"\n",
    "       c(\"all.rep (avg# 8652)\",  \"wo.rep*C (avg# 8435)\",\n",
    "         \"50%.wo.rep*C(avg# 4326)\",   \"10%.wo.rep*C(avg# 865)\",     \"10%.wo.rep*C(avg# 865)\",\n",
    "         \"2%.wo.rep*C(avg# 173)\"        \n",
    "                     #,\"\" #p(p(\"c=\",cuts[i]),\": spec.=\",round(cpfnc[2,i],3),\", sens.=\",round(cpfnc[1,i],3))\n",
    "                     ),lty=c(3,1,3,3,3,3),#,NA),\n",
    "       #cex=c(1,1,1,1,1,0.7),\n",
    "       col=c(\"blue\",\"black\",\"darkgreen\",\"magenta\",\"purple\",\"violet\"),#,\"black\"),\n",
    "       pch=c(1,3,4,5,5,6),#,NA),\n",
    "       cex=0.8, text.width=0.0032)\n",
    "cpfnc.sim<-cpfnc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cpfnc.def[,which(cuts%in%c(0.014,0.04,0.09))]\n",
    "cpfnc.sim[,which(cuts%in%c(0.014,0.04,0.09))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(rep.assign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(12,6,res=300)\n",
    "par(mfcol=c(1,2))\n",
    "list2env(cpf.similar.strains.eff.def.list,as.environment(-1))\n",
    "par(lab = c(10, 10, 7))\n",
    "plot(cpfnc[2,],cpfnc[1,],ylab=\"\",xlab=\"\",\n",
    "     type=\"b\",pch=3,lty=1,cex=0.8,col=\"black\",\n",
    "    xlim=c(0.99,1),ylim=c(0.55,1))\n",
    "#axis(2,at=seq(0,1,by=0.05))\n",
    "lines(cpff0.2s4[2,],cpff0.2s4[1,],type=\"b\",pch=5,lty=3,col=\"darkblue\")\n",
    "lines(cpff0.2[2,],cpff0.2[1,],type=\"b\",pch=5,lty=3,col=\"darkgreen\")\n",
    "lines(cpff0.2s1[2,],cpff0.2s1[1,],type=\"b\",pch=5,lty=3,col=\"magenta\")\n",
    "lines(cpff0.2s2[2,],cpff0.2s2[1,],type=\"b\",pch=5,lty=3,col=\"purple\")\n",
    "lines(cpff0.2s3[2,],cpff0.2s3[1,],type=\"b\",pch=5,lty=3,col=\"violet\")\n",
    "grid()\n",
    "i<-which(cuts%in%c(0.09,0.04,0.014))\n",
    "text(cpfnc[2,i],cpfnc[1,i],p(\"c=\",cuts[i]),col=\"red\",font=2,adj=c(0.5,0))\n",
    "i<-which(cuts%in%c(0.09))\n",
    "title(main=\"a)                        ROC curves for clustering cutoff                          \",line=0.5)\n",
    "title(ylab=\"sensitivity (true pairwise strain eq.)\",line=2.2); \n",
    "title(xlab=\"specificity (true pairwise strain ineq.)\",line=2.2)\n",
    "legend(\"bottomleft\",ncol=2,\n",
    "       c(#\"all.rep (avg# 8652)\",  \n",
    "         \"wo.rep*C (avg# 8435)\",\n",
    "         rep(\"20%wo.rep*C(avg# 1730)\",5)         \n",
    "         #rep(\"10%wo.rep*C(avg# 865)\",5)         \n",
    "                     #,\"\" #p(p(\"c=\",cuts[i]),\": spec.=\",round(cpfnc[2,i],3),\", sens.=\",round(cpfnc[1,i],3))\n",
    "                     ),lty=c(1,3,3,3,3,3),#,NA),\n",
    "       #cex=c(1,1,1,1,1,0.7),\n",
    "       col=c(\"black\",\"darkblue\",\"darkgreen\",\"magenta\",\"purple\",\"violet\"),#,\"black\"),\n",
    "       pch=c(3,5,5,5,5,5),#,NA),\n",
    "       cex=0.8, text.width=0.0032)\n",
    "\n",
    "list2env(cpf.similar.strains.eff.sim.list,as.environment(-1))\n",
    "par(lab = c(10, 10, 7))\n",
    "plot(cpfnc[2,],cpfnc[1,],ylab=\"\",xlab=\"\",type=\"b\",pch=3,lty=1,cex=0.8,col=\"black\",\n",
    "    xlim=c(0.99,1),ylim=c(0.55,1))\n",
    "#axis(2,at=seq(0,1,by=0.05))\n",
    "lines(cpff0.2s4[2,],cpff0.2s4[1,],type=\"b\",pch=5,lty=3,col=\"darkblue\")\n",
    "lines(cpff0.2[2,],cpff0.2[1,],type=\"b\",pch=5,lty=3,col=\"darkgreen\")\n",
    "lines(cpff0.2s1[2,],cpff0.2s1[1,],type=\"b\",pch=5,lty=3,col=\"magenta\")\n",
    "lines(cpff0.2s2[2,],cpff0.2s2[1,],type=\"b\",pch=5,lty=3,col=\"purple\")\n",
    "lines(cpff0.2s3[2,],cpff0.2s3[1,],type=\"b\",pch=5,lty=3,col=\"violet\")\n",
    "grid()\n",
    "i<-which(cuts%in%c(0.09,0.04,0.014))\n",
    "text(cpfnc[2,i],cpfnc[1,i],p(\"c=\",cuts[i]),col=\"red\",font=2,adj=c(0.5,0))\n",
    "i<-which(cuts%in%c(0.09))\n",
    "title(main=\"b)                        ROC curves for clustering cutoff                          \",line=0.5)\n",
    "#title(ylab=\"sensitivity\",line=2.2); title(xlab=\"specificity\",line=2.2)\n",
    "title(ylab=\"sensitivity (true pairwise strain eq.)\",line=2.2); \n",
    "title(xlab=\"specificity (true pairwise strain ineq.)\",line=2.2)\n",
    "legend(\"bottomleft\",ncol=2,\n",
    "       c(#\"all.rep (avg# 8652)\",  \n",
    "           \"wo.rep*C (avg# 8435)\",\n",
    "         rep(\"20%wo.rep*C(avg# 1730)\",5) \n",
    "         #rep(\"10%wo.rep*C(avg# 865)\",5)         \n",
    "                     #,\"\" #p(p(\"c=\",cuts[i]),\": spec.=\",round(cpfnc[2,i],3),\", sens.=\",round(cpfnc[1,i],3))\n",
    "                     ),lty=c(1,3,3,3,3,3),#,NA),\n",
    "       #cex=c(1,1,1,1,1,0.7),\n",
    "       col=c(\"black\",\"darkblue\",\"darkgreen\",\"magenta\",\"purple\",\"violet\"),#,\"black\"),\n",
    "       pch=c(3,5,5,5,5,5),#,NA),\n",
    "       cex=0.8, text.width=0.0032)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(6,6,res=300)\n",
    "list2env(cpf.similar.strains.def.list,as.environment(-1))\n",
    "plot(cpf[2,],cpf[1,],ylab=\"sensitivity\",xlab=\"specificity\",type=\"b\",pch=1,lty=3,cex=0.8,col=\"blue\",\n",
    "    xlim=c(0.7,1),ylim=c(0.7,1))\n",
    "#axis(2,at=seq(0,1,by=0.05))\n",
    "lines(cpfnc[2,],cpfnc[1,],type=\"b\",pch=3,lty=1)\n",
    "#lines(cpff0.5[2,],cpff0.5[1,],type=\"b\",pch=4,lty=3,col=\"darkgreen\")\n",
    "#lines(cpff0.1[2,],cpff0.1[1,],type=\"b\",pch=5,lty=3,col=\"magenta\")\n",
    "#lines(cpff0.1s2[2,],cpff0.1s2[1,],type=\"b\",pch=5,lty=3,col=\"purple\")\n",
    "lines(cpff0.02[2,],cpff0.02[1,],type=\"b\",pch=6,lty=3,col=\"violet\")\n",
    "grid(10)\n",
    "i<-which(cuts%in%c(0.09,0.07,0.13))\n",
    "text(cpfnc[2,i],cpfnc[1,i],p(\"c=\",cuts[i]),col=\"red\",font=2,adj=c(0.5,0))\n",
    "i<-which(cuts%in%c(0.09))\n",
    "title(main=\"ROC curve for clustering cutoff\")\n",
    "legend(\"bottomleft\",ncol=2,\n",
    "       c(\"all.rep\",\"wo.rep*C\",#\"50%.data.wo.rep*C\",\"10%.data.wo.rep*C\",\"10%.data.wo.rep*C\",\n",
    "         \"2%.data.wo.rep*C\"\n",
    "                     #,\"\" #p(p(\"c=\",cuts[i]),\": spec.=\",round(cpfnc[2,i],3),\", sens.=\",round(cpfnc[1,i],3))\n",
    "                     ),lty=c(3,1,3),#3,3,3),#,NA),\n",
    "       #cex=c(1,1,1,1,1,0.7),\n",
    "       col=c(\"blue\",\"black\",\"violet\"),#\"magenta\",\"purple\",\"violet\"),#,\"black\"),\n",
    "       pch=c(1,3,5),#5,5,6),#,NA),\n",
    "       text.width=0.10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cpff0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#colnames(samp.len.cnts.ma.norm)\n",
    "head(rep.assign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "toinsp.names\n",
    "head(rep.assign)\n",
    "sum(rep.assign$time==\"A\")+sum(rep.assign$time==\"B\")+sum(rep.assign$time==\"C\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(8,res=500)\n",
    "options(repr.plot.res=300)\n",
    "#options(repr.plot.width=10, repr.plot.height=5)\n",
    "#TODO inspect repBC94A, repBC87A, repBC89C why two first are more similar to each other than last two\n",
    "#\n",
    "#sel.species<-which(grepl(\"repBC96B|repBC95B|repBC31B\",colnames(samp.len.cnts.ma.norm)))\n",
    "\n",
    "sel.species<-which(grepl(\"repBC94A|repBC87A|repBC89C\",colnames(samp.len.cnts.ma.norm)))\n",
    "#sel.species<-match(c(\"repBC94A\",\"repBC87A\",\"repBC89C\"),colnames(samp.len.cnts.ma.norm))\n",
    "sel.species<-c(#toinsp.names[4:6]\n",
    "             #  \"repBC28C\",\n",
    "             #  \"repBC29B\", \"repBC28A\" #\"repBC28B\"\n",
    "             #\"repBC94A\", \"repBC86A\", \"repBC95B\"\n",
    "             \"repBC87A\",\"repBC88B\",\"repBC81C\"#\"repBC96B\"\n",
    "              ) \n",
    "               #\"repBC08A\") \n",
    "                #match(c(\"repBC28C\",\"repBC29B\",\"repBC08A\"),colnames(samp.len.cnts.ma.norm))\n",
    "sel.species #colnames(samp.len.cnts.ma.norm)[sel.species]\n",
    "with(rep.assign,rep.assign[match(sel.species,rep.name),\"rep.name\"])\n",
    "with(rep.assign,rep.assign[match(sel.species,rep.name),\"full.strain\"])\n",
    "\n",
    "all(cbind(colnames(samp.len.cnts.ma.norm)==as.character(rep.assign$rep.name)))\n",
    "\n",
    "#toplot.mat\n",
    "\n",
    "#alpha<-c(0.05,0.1,0.15,0.2,0.4)\n",
    "\n",
    "par(mfrow=c(1+length(sel.species),1),mai = c(0.8, 0.8, 0.5, 0.1),oma=c(0,0,0,0))\n",
    "layout(matrix(c(1:4),4, 1, byrow = TRUE), widths=c(1,1,1,1), heights=c(1.21,0.93,0.93,0.93))\n",
    "\n",
    "#plot(cnt.range,h0.sample[cnt.range],type=\"l\",pch=\".\",xlab=\"h0.sample\")\n",
    "to.prob<-function(v)v/sum(v)\n",
    "plot(cnt.range,to.prob(rowSums(samp.len.cnts[cnt.range,rep.assign$time==\"A\"])),type=\"l\",pch=\".\",\n",
    "     col=\"darkred\",\n",
    "     xlab=\"read length\",ylab=\"All reads density\")\n",
    "lines(cnt.range,to.prob(rowSums(samp.len.cnts[cnt.range,rep.assign$time==\"B\"])),type=\"l\",pch=\".\",\n",
    "     col=\"darkblue\")\n",
    "lines(cnt.range,to.prob(rowSums(samp.len.cnts[cnt.range,rep.assign$time==\"C\"])),type=\"l\",pch=\".\",\n",
    "     col=\"darkgreen\")\n",
    "legend(\"topright\",#title=paste(\"Reads gathered\\nin sequencing runs A,B,C\"),\n",
    "       legend=c(paste(\"Distribution of reads gathered\\n     in sequencing run:\"), \"A\",\"B\",\"C\"),lty=c(NA,1,1,1),col=c(\"black\",\"darkred\",\"darkblue\",\"darkgreen\"),y.intersp=0.02,cex=0.8)\n",
    "\n",
    "lv<-NULL\n",
    "sel.vecs<-mapply(function(i,j){\n",
    "    par(mai = c(0.8, 0.8, 0.1, 0.1))\n",
    "    #v<-samp.len.cnts.norm[cnt.range,i] \n",
    "    v<-samp.len.cnts.ma.norm[cnt.range,i]\n",
    "    #v<-log10(0.00001+v)\n",
    "    xlab<-paste(i,rep.assign.cn[i,\"full.strain\"])\n",
    "    #legend(\"topright\",legend=paste(i,\" #reads:\",sum(samp.len.cnts[,i])))\n",
    "    \n",
    "    lv<-10^(1.0*log.samp.prob[,i])\n",
    "    \n",
    "    #Correct short reads\n",
    "    \n",
    "    rc<-sum(samp.len.cnts[cnt.range,i])\n",
    "    rratio<-10^((5000-rc)*tail(coef(res.short.lm),1)) #log10 regression model\n",
    "    s<-sum(samp.len.cnts[cnt.range.short,i]*cnt.range.short.w)/rc\n",
    "    p<-rratio*s*(1-s)/(1-rratio*s*s)\n",
    "    \n",
    "    #lv[cnt.range.short]<-lv[cnt.range.short]/p*cnt.range.short.w\n",
    "    \n",
    "    \n",
    "    plot(x=cnt.range,lv,type=\"l\",pch=\".\",xlab=xlab,ylab=\"reads density\")#,ylim=c(-6,-2))#colnames(samp.len.cnts)[i])    \n",
    "    \n",
    "    legend(\"topright\",legend=paste(i,\" #reads:\",sum(samp.len.cnts[cnt.range,i]),\n",
    "                \"\\n mean.read.length\", \n",
    "                round(sum(samp.len.cnts[cnt.range,i]*cnt.range)/sum(samp.len.cnts[cnt.range,i])),1)\n",
    "    )\n",
    "    \n",
    "    #print(sum(lv))\n",
    "    #lv<-log.samp.prob[,i]\n",
    "    print(summary(lv))\n",
    "    print(min.stats(lv,3))\n",
    "    #lv<-lv/sum(lv)\n",
    "    #lv<-log10(lv)\n",
    "    points(x=cnt.range,v,type=\"l\",pch=\".\",col=c(\"darkred\",\"darkblue\",\"darkgreen\")[j],xlab=i) #colnames(samp.len.cnts)[i],col=\"red\")\n",
    "    v\n",
    "},sel.species,seq_along(sel.species))#,alpha,round(alpha*50))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min.stats<-function(v,k=3,...){\n",
    "    if (k==0) return(NULL)\n",
    "    mv<-min(v,...)\n",
    "    idx<-which(v==mv)\n",
    "    nv<-v; nv[idx]<-Inf\n",
    "    rbind(c(min=mv,p=length(idx)/length(v)),\n",
    "          min.stats(nv,k=k-1,...))\n",
    "}\n",
    "\n",
    "\n",
    "min.stats(log.samp.prob[,20],3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "psz(12)\n",
    "#options(repr.plot.width=10, repr.plot.height=5)\n",
    "#TODO inspect repBC94A, repBC87A, repBC89C why two first are more similar to each other than last two\n",
    "#\n",
    "#sel.species<-which(grepl(\"repBC96B|repBC95B|repBC31B\",colnames(samp.len.cnts.ma.norm)))\n",
    "\n",
    "sel.species<-match(c(\"repBC04B\",\"repBC11A\",\"repBC06C\",\"repBC04A\",\"repBC04C\",\"repBC05C\"),colnames(samp.len.cnts.ma.norm))\n",
    "colnames(samp.len.cnts.ma.norm)[sel.species]\n",
    "rep.assign[sel.species,\"rep.name\"]\n",
    "rep.assign[sel.species,\"full.strain\"]\n",
    "\n",
    "all(cbind(colnames(samp.len.cnts.ma.norm)==as.character(rep.assign$rep.name)))\n",
    "\n",
    "#toplot.mat\n",
    "\n",
    "#alpha<-c(0.05,0.1,0.15,0.2,0.4)\n",
    "\n",
    "par(mfrow=c((1+length(sel.species))/2+1,2))\n",
    "plot(cnt.range,h0.sample[cnt.range],type=\"l\",pch=\".\",xlab=\"h0.sample\")\n",
    "plot(1:100)\n",
    "sel.vecs<-mapply(function(i,j){\n",
    "    v<-ma(samp.len.cnts.norm[,i],5)[cnt.range]\n",
    "    v<-log10(10^-5+v)\n",
    "    xlab<-paste(colnames(samp.len.cnts)[i],rep.assign[i,\"full.strain\"])\n",
    "    plot(x=cnt.range,v,type=\"l\",pch=\".\",xlab=xlab,ylim=c(-6,-2))\n",
    "    legend(\"topright\",legend=paste(i,\" #reads:\",sum(samp.len.cnts[cnt.range,i]),\n",
    "                \"\\n mean.read.length\", \n",
    "                round(sum(samp.len.cnts[cnt.range,i]*cnt.range)/sum(samp.len.cnts[cnt.range,i])),1)\n",
    "    )\n",
    "    lv<-10^(0.5*log.samp.prob[,2+i])\n",
    "    lv<-lv/sum(lv)\n",
    "    lv<-log.samp.prob[,2+i]\n",
    "    points(x=cnt.range,lv,type=\"l\",pch=\".\",xlab=colnames(samp.len.cnts)[i],col=\"red\")\n",
    "    v\n",
    "},sel.species,seq_along(sel.species))#,alpha,round(alpha*50))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "odds.ratio.norm.fun<-Vectorize(function(i,j){\n",
    "#i<-i_j[1];j<-i_j[2]\n",
    "lp<-log.samp.prob[,2+j]\n",
    "lp0<-log.samp.prob[,1]\n",
    "p<-0.5\n",
    "lp<-log10(10^(p*lp)/sum(10^(p*lp)))\n",
    "lp0<-log10(10^(p*lp0)/sum(10^(p*lp0)))\n",
    "sum(lp*samp.len.cnts[cnt.range,i])/sum(samp.len.cnts[cnt.range,i])-sum(lp0*samp.len.cnts[cnt.range,i])/sum(samp.len.cnts[cnt.range,i])\n",
    "#dist.diff<-(cumsum(samp.len.cnts[cnt.range,j])/sum(samp.len.cnts[cnt.range,j]) -\n",
    "#        cumsum(samp.len.cnts[cnt.range,i])/sum(samp.len.cnts[cnt.range,i]))\n",
    "#print(which.max(abs(dist.diff)))\n",
    "#max(abs(dist.diff))\n",
    "})\n",
    "#outer(sel.species,sel.species,FUN=function(x,y)paste(x,y))\n",
    "m<-matrix(NA,length(sel.species),length(sel.species))\n",
    "colnames(m)<-sel.species\n",
    "rownames(m)<-sel.species\n",
    "m[]<-odds.ratio.norm.fun(sel.species[row(m)],sel.species[col(m)])\n",
    "m\n",
    "10^((m+t(m))/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Regression on read count analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#head(rep.assign)\n",
    "#table(factor(rep.assign$full.strain))\n",
    "#hist(colSums(samp.len.cnts))\n",
    "rep.set1<-as.character(rep.assign[as.integer(factor(rep.assign$full.strain))==1,\"rep.name\"])\n",
    "rep.set1\n",
    "range(cnt.range)\n",
    "#samp.len.cnts.norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all(rep.assign$rep.name==colnames(samp.len.cnts))\n",
    "dim(samp.len.cnts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Linear regression\n",
    "# Design matrix construction\n",
    "fs.fac<-as.integer(factor(rep.assign$.strain.))\n",
    "X<-matrix(0,ncol=max(fs.fac),nrow=NROW(rep.assign))\n",
    "X[col(X)==fs.fac&row(X)==seq(along.with=fs.fac)]<-1\n",
    "\n",
    "samp.read.counts.cnt.range<-colSums(samp.len.cnts[cnt.range,])\n",
    "#samp.read.counts.cnt.range<-colSums(samp.len.cnts.ma.norm[cnt.range,])\n",
    "\n",
    "samp.mean.read.length.cnt.range<-colSums(samp.len.cnts[cnt.range,]*cnt.range)/samp.read.counts.cnt.range\n",
    "Y<-samp.mean.read.length.cnt.range\n",
    "Y<-colSums(samp.len.cnts.ma.norm[cnt.range,]*cnt.range)/colSums(samp.len.cnts.ma.norm[cnt.range,])\n",
    "\n",
    "\n",
    "psz(8,4)\n",
    "par(mfcol=c(1,2))\n",
    "hist(Y,breaks=35)\n",
    "hist(samp.read.counts.cnt.range,breaks=35)\n",
    "xyplot(Y~samp.read.counts.cnt.range, type=c(\"p\",\"fit\"))\n",
    "#X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(cnt.range)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Y == samp.mean.read.length.cnt.range\n",
    "#res.lm<-lm(Y~X+samp.read.counts.cnt.range+0+X*samp.read.counts.cnt.range)\n",
    "res.lm<-lm(Y~X+samp.read.counts.cnt.range+0)\n",
    "summary(res.lm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "v<-seq(300,3000,by=10)\n",
    "names(v)<-v\n",
    "short.lm.coefs<-sapply(v,function(r){\n",
    "cnt.range.short<-cnt.range[1]:r #180:r \n",
    "Yshort<-colSums(samp.len.cnts[cnt.range.short,])/samp.read.counts.cnt.range\n",
    "Yshort<-Yshort/sqrt(sum(Yshort^2))\n",
    "Yshort<-Yshort-mean(Yshort)\n",
    "res.short.lm<-lm(Yshort~X+samp.read.counts.cnt.range+0)\n",
    "cbind(tail(coef(res.short.lm),1),summary(res.short.lm)$r.squared)\n",
    "})\n",
    "#summary(res.short.lm)\n",
    "\n",
    "#cnt.range.short<-cnt.range[1]:as.integer(names(v)[which.max(short.lm.coefs)])\n",
    "cnt.range.short<-c(cnt.range[1]:890,891:1000)\n",
    "cnt.range.short.w<-c(rep(1,890-cnt.range[1]+1),seq(1,0,length.out=1000-890))\n",
    "Yshort<-colSums(samp.len.cnts[cnt.range.short,]*cnt.range.short.w)/samp.read.counts.cnt.range\n",
    "res.short.lm<-lm(Yshort~X+samp.read.counts.cnt.range+0)\n",
    "summary(res.short.lm)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "short.lm.coefs[1,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(t(short.lm.coefs))\n",
    "short.lm.coefs[,which.max(short.lm.coefs[1,]),drop=F]\n",
    "short.lm.coefs[,which.max(short.lm.coefs[2,]),drop=F]\n",
    "psz(5)\n",
    "plot(x=as.integer(names(v)),y=short.lm.coefs[1,],type=\"b\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tail(coef(res.short.lm),1)\n",
    "slm<-summary(res.short.lm)\n",
    "slm$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tail(coef(res.lm),1)\n",
    "tail(coef(res.lm),1)*10000\n",
    "#grepl(\"Salmonella\",rep.assign$full.strain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "YhatConst<-X%*%head(coef(res.lm),-1)\n",
    "Yhat<-YhatConst+samp.read.counts.cnt.range*tail(coef(res.lm),1)\n",
    "YshorthatConst<-X%*%head(coef(res.short.lm),-1)\n",
    "Yshorthat<-YshorthatConst+samp.read.counts.cnt.range*tail(coef(res.short.lm),1)\n",
    "\n",
    "res.lm.df<-data.frame(samp.mean.read.length.cnt.range=Y,\n",
    "                      Yhat=Yhat,\n",
    "                      Yshort=Yshort, \n",
    "                      samp.read.counts.cnt.range=samp.read.counts.cnt.range,\n",
    "                      full.strain=rep.assign$full.strain)\n",
    "psz(10,10, res=150)\n",
    "xyplot(samp.mean.read.length.cnt.range #+Yhat\n",
    "       ~samp.read.counts.cnt.range|full.strain,\n",
    "       data=res.lm.df,\n",
    "       #subset=as.integer(factor(rep.assign$full.strain))%in%c(1:60),\n",
    "       #subset=grepl(\"Bacillus_cereus_38|Salmonella\",\n",
    "       #             rep.assign$full.strain),\n",
    "       par.strip.text=list(cex=0.5),\n",
    "      panel = function(x,y,subscripts,...) {\n",
    "            #print(paste(subscripts))            \n",
    "            panel.lmline(x, y,  ..., col=\"blue\",lty=3)\n",
    "            panel.abline(b=tail(coef(res.lm),1), a=YhatConst[subscripts[1]],col=\"red4\",lty=5)            \n",
    "            to.mark<-rep.assign[subscripts,\"rep.name\"]%in%c(\"repBC87A\",\"repBC81C\") #c(\"repBC88B\",\"repBC81C\")\n",
    "            panel.xyplot(x, y, ...)              \n",
    "            panel.points(x[to.mark],y[to.mark],pch=3,col=\"green3\",cex=2)\n",
    "            panel.text(x,y,rep.assign$rep.name[subscripts],cex=0.5)            \n",
    "      },\n",
    "      xlab=\"sample.read.count\",\n",
    "      ylab=\"sample.mean.read.length\",\n",
    "      )\n",
    "\n",
    "psz(10,10, res=150)\n",
    "xyplot(log10(Yshort) #+Yhat\n",
    "       ~samp.read.counts.cnt.range|full.strain,\n",
    "       data=res.lm.df,\n",
    "       subset=as.integer(factor(rep.assign$full.strain))%in%c(1:20),\n",
    "       #subset=grepl(\"Bacillus_cereus_38|Salmonella\",\n",
    "       #             rep.assign$full.strain),\n",
    "       par.strip.text=list(cex=0.5),\n",
    "      panel = function(x,y,subscripts,...) {\n",
    "            #print(paste(subscripts))            \n",
    "            panel.lmline(x, y,  ..., col=\"blue\",lty=3)\n",
    "            panel.abline(b=tail(coef(res.short.lm),1), a=YshorthatConst[subscripts[1]],col=\"red4\",lty=5)            \n",
    "            to.mark<-rep.assign[subscripts,\"rep.name\"]%in%c(\"repBC87A\",\"repBC81C\") #c(\"repBC88B\",\"repBC81C\")\n",
    "            panel.xyplot(x, y, ...)              \n",
    "            panel.points(x[to.mark],y[to.mark],pch=3,col=\"green3\",cex=2)\n",
    "            panel.text(x,y,rep.assign$rep.name[subscripts],cex=0.5)            \n",
    "      },\n",
    "      xlab=\"sample.read.count\",\n",
    "      #ylab=\"sample.mean.read.length\",\n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cs<-colSums(samp.len.cnts[cnt.range,rep.set1])\n",
    "\n",
    "df<-samp.len.cnts.ma.norm[cnt.range,rep.set1]-rowMeans(samp.len.cnts.ma.norm[cnt.range,rep.set1])\n",
    "\n",
    "df[5]\n",
    "\n",
    "psz(w=10,h=4)\n",
    "for (i in seq_along(rep.set1)){\n",
    "plot(cnt.range,\n",
    "    df[,i],main=cs[i],\n",
    "    type = 'l'\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=5)\n",
    "\n",
    "sel.species<-c(1,5,10,15,20)\n",
    "alpha<-c(0.05,0.1,0.15,0.2,0.4)\n",
    "\n",
    "par(mfrow=c(3,2))\n",
    "sel.vecs<-mapply(function(i,j,a,n){\n",
    "    v<-samp.len.cnts.ma.norm[1:2500,i]*a*100    \n",
    "    plot(v,type=\"l\",pch=\".\",xlab=p(\"a\",j,\"=\",alpha[j]))\n",
    "    v\n",
    "},sel.species,seq_along(sel.species),alpha,round(alpha*50))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=3)\n",
    "plot(rowSums(sel.vecs),type=\"l\",pch=\".\",xlab=\"read.length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prob<-rowSums(sel.vecs)\n",
    "prob[is.na(prob)]<-0\n",
    "Nreads<-2000\n",
    "sel.sample <- sample(seq_len(NROW(sel.vecs)),Nreads,prob=prob,replace = TRUE)\n",
    "sel.samp.hist<-hist(sel.sample,xlim=c(1,2500),breaks=100,xlab=\"read.length\",sub=p(\"Nreads=\",Nreads))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#######"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Save .RData to project dir\n",
    "gc()\n",
    "system.time(save_new_image())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.2"
  },
  "name": "__NB_TITLE",
  "title": "__NB_TITLE"
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
